{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#アミノ酸配列、ICTVによるリファレンスあり\" data-toc-modified-id=\"アミノ酸配列、ICTVによるリファレンスあり-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>アミノ酸配列、ICTVによるリファレンスあり</a></span><ul class=\"toc-item\"><li><span><a href=\"#リファレンスツリーの配列数のカウント\" data-toc-modified-id=\"リファレンスツリーの配列数のカウント-1.1\"><span class=\"toc-item-num\">1.1&nbsp;&nbsp;</span>リファレンスツリーの配列数のカウント</a></span></li><li><span><a href=\"#ラベルデータの読み込み（作成：Final_dataset_phylogenetic_tree_label）\" data-toc-modified-id=\"ラベルデータの読み込み（作成：Final_dataset_phylogenetic_tree_label）-1.2\"><span class=\"toc-item-num\">1.2&nbsp;&nbsp;</span>ラベルデータの読み込み（作成：Final_dataset_phylogenetic_tree_label）</a></span><ul class=\"toc-item\"><li><span><a href=\"#known-viruses-vs-unknown-viruses\" data-toc-modified-id=\"known-viruses-vs-unknown-viruses-1.2.1\"><span class=\"toc-item-num\">1.2.1&nbsp;&nbsp;</span>known viruses vs unknown viruses</a></span></li></ul></li><li><span><a href=\"#blastxによって、該当アミノ酸領域を塩基配列から取り出す\" data-toc-modified-id=\"blastxによって、該当アミノ酸領域を塩基配列から取り出す-1.3\"><span class=\"toc-item-num\">1.3&nbsp;&nbsp;</span>blastxによって、該当アミノ酸領域を塩基配列から取り出す</a></span><ul class=\"toc-item\"><li><span><a href=\"#blastxの結果から配列取得\" data-toc-modified-id=\"blastxの結果から配列取得-1.3.1\"><span class=\"toc-item-num\">1.3.1&nbsp;&nbsp;</span>blastxの結果から配列取得</a></span></li><li><span><a href=\"#bitscoreの最も高い行を1つずつ、各subjectに対して選択\" data-toc-modified-id=\"bitscoreの最も高い行を1つずつ、各subjectに対して選択-1.3.2\"><span class=\"toc-item-num\">1.3.2&nbsp;&nbsp;</span>bitscoreの最も高い行を1つずつ、各subjectに対して選択</a></span></li><li><span><a href=\"#blastxでヒットしなかった配列の数をチェック（virus_IDとしてどの程度減ったか？）\" data-toc-modified-id=\"blastxでヒットしなかった配列の数をチェック（virus_IDとしてどの程度減ったか？）-1.3.3\"><span class=\"toc-item-num\">1.3.3&nbsp;&nbsp;</span>blastxでヒットしなかった配列の数をチェック（virus_IDとしてどの程度減ったか？）</a></span></li><li><span><a href=\"#Final_dataset_phylogenetic_tree_label.ipynbで作成したラベル表に入っていない配列は除去する\" data-toc-modified-id=\"Final_dataset_phylogenetic_tree_label.ipynbで作成したラベル表に入っていない配列は除去する-1.3.4\"><span class=\"toc-item-num\">1.3.4&nbsp;&nbsp;</span>Final_dataset_phylogenetic_tree_label.ipynbで作成したラベル表に入っていない配列は除去する</a></span><ul class=\"toc-item\"><li><span><a href=\"#（予測タスクが行われていないという意味）\" data-toc-modified-id=\"（予測タスクが行われていないという意味）-1.3.4.1\"><span class=\"toc-item-num\">1.3.4.1&nbsp;&nbsp;</span>（予測タスクが行われていないという意味）</a></span></li></ul></li><li><span><a href=\"#virus_IDがデータセット間で共通の場合が存在？\" data-toc-modified-id=\"virus_IDがデータセット間で共通の場合が存在？-1.3.5\"><span class=\"toc-item-num\">1.3.5&nbsp;&nbsp;</span>virus_IDがデータセット間で共通の場合が存在？</a></span><ul class=\"toc-item\"><li><span><a href=\"#sars-cov-2データセットの中に、coronaviridae-unknownと冗長なものが含まれていた\" data-toc-modified-id=\"sars-cov-2データセットの中に、coronaviridae-unknownと冗長なものが含まれていた-1.3.5.1\"><span class=\"toc-item-num\">1.3.5.1&nbsp;&nbsp;</span>sars-cov-2データセットの中に、coronaviridae unknownと冗長なものが含まれていた</a></span></li><li><span><a href=\"#例外対応はCoronaviridaeのみ：influenzaの場合はこの心配はない\" data-toc-modified-id=\"例外対応はCoronaviridaeのみ：influenzaの場合はこの心配はない-1.3.5.2\"><span class=\"toc-item-num\">1.3.5.2&nbsp;&nbsp;</span>例外対応はCoronaviridaeのみ：influenzaの場合はこの心配はない</a></span></li></ul></li><li><span><a href=\"#known-vs-unknown-viruses\" data-toc-modified-id=\"known-vs-unknown-viruses-1.3.6\"><span class=\"toc-item-num\">1.3.6&nbsp;&nbsp;</span>known vs unknown viruses</a></span></li><li><span><a href=\"#それぞれのモデルで予測確率がnp.nanである配列の数\" data-toc-modified-id=\"それぞれのモデルで予測確率がnp.nanである配列の数-1.3.7\"><span class=\"toc-item-num\">1.3.7&nbsp;&nbsp;</span>それぞれのモデルで予測確率がnp.nanである配列の数</a></span></li></ul></li><li><span><a href=\"#【出力】ラベル\" data-toc-modified-id=\"【出力】ラベル-1.4\"><span class=\"toc-item-num\">1.4&nbsp;&nbsp;</span>【出力】ラベル</a></span></li><li><span><a href=\"#blastxの結果からアミノ酸配列を書き出し\" data-toc-modified-id=\"blastxの結果からアミノ酸配列を書き出し-1.5\"><span class=\"toc-item-num\">1.5&nbsp;&nbsp;</span>blastxの結果からアミノ酸配列を書き出し</a></span></li><li><span><a href=\"#mafftでのアライメント-+-リファレンス配列の削除\" data-toc-modified-id=\"mafftでのアライメント-+-リファレンス配列の削除-1.6\"><span class=\"toc-item-num\">1.6&nbsp;&nbsp;</span>mafftでのアライメント + リファレンス配列の削除</a></span></li><li><span><a href=\"#iqtreeで系統樹作成\" data-toc-modified-id=\"iqtreeで系統樹作成-1.7\"><span class=\"toc-item-num\">1.7&nbsp;&nbsp;</span>iqtreeで系統樹作成</a></span></li></ul></li></ul></div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-23T06:36:38.227040Z",
     "start_time": "2024-03-23T06:36:38.175297Z"
    }
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/env python\n",
    "import sys, re\n",
    "argvs = sys.argv\n",
    "sys.setrecursionlimit(10000000)\n",
    "\n",
    "import os\n",
    "import collections\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import array\n",
    "pd.set_option(\"display.max_colwidth\", 200)\n",
    "pd.set_option(\"display.max_columns\", 200)\n",
    "pd.set_option(\"display.max_rows\", 300)\n",
    "\n",
    "import glob\n",
    "\n",
    "## visualize\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "sns.set()\n",
    "\n",
    "plt.rcParams['font.family'] = \"Arial\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Using reference alignment (amino acid sequences) published by ICTV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-23T06:36:39.165055Z",
     "start_time": "2024-03-23T06:36:39.161609Z"
    }
   },
   "outputs": [],
   "source": [
    "virus=\"Coronaviridae\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-23T06:36:39.863541Z",
     "start_time": "2024-03-23T06:36:39.858548Z"
    }
   },
   "outputs": [],
   "source": [
    "if virus == \"Coronaviridae\":\n",
    "    virus_1 = \"sarbecovirus\"\n",
    "    virus_l = [virus, virus_1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Count sequence number in the reference alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-23T06:36:40.728781Z",
     "start_time": "2024-03-23T06:36:40.724212Z"
    }
   },
   "outputs": [],
   "source": [
    "def count_signs(filename):\n",
    "    count = 0\n",
    "    with open(filename, 'r') as file:\n",
    "        for line in file:\n",
    "            if '>' in line:\n",
    "                count += 1\n",
    "    return count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-23T06:36:40.989428Z",
     "start_time": "2024-03-23T06:36:40.983018Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "55\n"
     ]
    }
   ],
   "source": [
    "tree_fasta_f = \"/\".join(map(str, [\"input_examples\", \"Coronaviridae\", \"ictv.fasta\"]))\n",
    "print (count_signs(tree_fasta_f))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## convert nucleotide sequences to protein sequences using tBLASTn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-23T06:40:31.516138Z",
     "start_time": "2024-03-23T06:39:04.645369Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 05/02/2024 16:45:34\n",
      "New DB name:   /Users/junna/Desktop/research/_2024/001_VirusBERT/000_git-hub/BERT-infect_2024/007_phylogenetic_analysis/output_examples/Coronaviridae/add.DB\n",
      "New DB title:  output_examples/Coronaviridae/add.fasta\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 169 sequences in 0.0403759 seconds.\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: output_examples/Coronaviridae: File exists\n",
      "[INFO]\u001b[0m 12689 patterns loaded from file\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "train_virus=\"Coronaviridae\"\n",
    "pred_virus=\"sarbecov\"\n",
    "\n",
    "mkdir output_examples/${virus}\n",
    "        \n",
    "## リファレンス配列からギャップを除く\n",
    "cat input_examples/${virus}/ictv.fasta | \\\n",
    "    seqkit seq -g \\\n",
    "    > output_examples/${virus}/ictv.ref.fasta\n",
    "\n",
    "## 配列ファイルの指定\n",
    "fasta_f=../002_Data_splitting/input_examples/${virus}.curated.fasta\n",
    "        \n",
    "## Accession Noの取り出し\n",
    "cat input_examples/final_dataset_phylogenetic_tree_label.txt | \\\n",
    "    grep -e ${train_virus} -e ${pred_virus} | awk 'BEGIN{FS=\"\\t\"}NR>2{print $1}' \\\n",
    "    > output_examples/${virus}/accession.txt\n",
    "\n",
    "## 配列取り出し\n",
    "cat ${fasta_f} | \\\n",
    "    seqkit seq -g | \\\n",
    "    seqkit grep -f output_examples/${virus}/accession.txt \\\n",
    "    > output_examples/${virus}/add.fasta\n",
    "\n",
    "## blast DB作成\n",
    "makeblastdb \\\n",
    "    -in output_examples/${virus}/add.fasta \\\n",
    "    -out output_examples/${virus}/add.DB \\\n",
    "    -dbtype nucl\n",
    "\n",
    "## blastp\n",
    "tblastn \\\n",
    "    -query output_examples/${virus}/ictv.ref.fasta \\\n",
    "    -db output_examples/${virus}/add.DB \\\n",
    "    -out  output_examples/${virus}/tblastn.txt \\\n",
    "    -outfmt '6 qseqid sseqid evalue bitscore length qlen slen pident sseq' "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-23T06:48:12.249915Z",
     "start_time": "2024-03-23T06:48:12.085006Z"
    }
   },
   "outputs": [],
   "source": [
    "virus = virus_l[0]\n",
    "\n",
    "## read blast result\n",
    "blast_l = []\n",
    "\n",
    "blast_f = \"/\".join(map(str, [\"output_examples\", virus , \"tblastn.txt\"]))\n",
    "df_blast = pd.read_csv(blast_f, sep=\"\\t\", \n",
    "                       names=[\"query\", \"subject\", \"evalue\", \"bitscore\", \"alen\", \"qlen\", \"slen\", \"pident\", \"sseq\"])\n",
    "    \n",
    "df_blast = df_blast.astype({\"alen\":int, \"qlen\": int, \"bitscore\":float})\n",
    "df_blast[\"coverage\"] = df_blast[\"alen\"] / df_blast[\"qlen\"] * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-23T06:48:15.629069Z",
     "start_time": "2024-03-23T06:48:15.513804Z"
    }
   },
   "outputs": [],
   "source": [
    "# select tophit sequence\n",
    "idx = df_blast.groupby('subject')['bitscore'].idxmax()\n",
    "result = df_blast.loc[idx]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-23T06:48:49.688342Z",
     "start_time": "2024-03-23T06:48:49.578636Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output_examples/Coronaviridae/tblastn.fasta\n"
     ]
    }
   ],
   "source": [
    "## output protein sequence\n",
    "\n",
    "# read virus_ID\n",
    "df_tree = pd.read_csv(\"input_examples/final_dataset_phylogenetic_tree_label.txt\",\n",
    "                      sep=\"\\t\")\n",
    "output_l = df_tree[\"virus_ID\"].unique()\n",
    "df_subject = result[result[\"subject\"].isin(output_l)]\n",
    "df_subject = df_subject.loc[:, [\"subject\", \"sseq\"]]\n",
    "\n",
    "# write protein sequences\n",
    "output_f =  \"/\".join(map(str, [\"output_examples\", virus_l[0], \"tblastn.fasta\"]))\n",
    "print (output_f)\n",
    "\n",
    "f = open(output_f, \"w\")\n",
    "for i, row in df_subject.iterrows():\n",
    "    header, seq = row[:2]\n",
    "    f.write(\">\" + header + \"\\n\")\n",
    "    f.write(seq + \"\\n\")\n",
    "\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# mafft alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-23T06:52:55.508126Z",
     "start_time": "2024-03-23T06:49:10.059811Z"
    },
    "run_control": {
     "marked": true
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "outputhat23=16\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8176 kb\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "tbfast-pair (aa) Version 7.508\n",
      "alg=L, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "outputhat23=16\n",
      "Loading 'hat3.seed' ... \n",
      "done.\n",
      "Writing hat3 for iterative refinement\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 1, compacttree = 0\n",
      "Constructing a UPGMA tree ... \n",
      "   50 / 55\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    54 /54 \n",
      "done.\n",
      "tbfast (aa) Version 7.508\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 16\n",
      "sueff_global = 0.100000\n",
      "nadd = 16\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "   50 / 55\n",
      "Segment   1/  1    1- 317\n",
      "STEP 003-001-0  identical.   \n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (aa) Version 7.508\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " L-INS-i (Probably most accurate, very slow)\n",
      " Iterative refinement method (<16) with LOCAL pairwise alignment information\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "nadd =  169\n",
      "npair =  9295\n",
      "nseq =  224\n",
      "nlen =  315\n",
      "use multipair, weighti=2.7!\n",
      "nadd = 169\n",
      "rescale = 1\n",
      "    0 / 55 (thread    0)dndpre (aa) Version 7.508\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "4 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.508\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "4 thread(s)\n",
      "\n",
      "nadd = 169\n",
      "ppenalty_ex = -10\n",
      "nthread = 4\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 55\n",
      "njobc = 56\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "cTEP 168 / 169 (thread 2)                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.508\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, +0.01, noshift, amax=0.0\n",
      "4 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-fragment (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "[INFO]\u001b[0m 55 patterns loaded from file\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "train_virus=\"Coronaviridae\"\n",
    "        \n",
    "## remove gaps from reference alignment\n",
    "cat input_examples/${train_virus}/ictv.fasta | \\\n",
    "    seqkit seq -g \\\n",
    "    > output_examples/${train_virus}/ictv.ref.fasta\n",
    "        \n",
    "## mafft\n",
    "mafft --auto \\\n",
    "    output_examples/${train_virus}/ictv.ref.fasta \\\n",
    "    > output_examples/${train_virus}/ictv.ref.mafft.fasta\n",
    "        \n",
    "## mafft -add\n",
    "mafft --auto --keeplength \\\n",
    "    --addfragments output_examples/${train_virus}/tblastn.fasta \\\n",
    "    --thread 4 \\\n",
    "    output_examples/${train_virus}/ictv.ref.mafft.fasta \\\n",
    "    > output_examples/${train_virus}/tblastn.mafft.fasta\n",
    "        \n",
    "## extract reference sequence names\n",
    "cat output_examples/${train_virus}/ictv.ref.fasta | \\\n",
    "    grep \">\" | sed 's/>//g' \\\n",
    "    > output_examples/${train_virus}/ictv.ref.txt\n",
    "        \n",
    "## remove reference sequences\n",
    "cat output_examples/${train_virus}/tblastn.mafft.fasta | \\\n",
    "    seqkit grep -v -f output_examples/${train_virus}/ictv.ref.txt \\\n",
    "    > output_examples/${train_virus}/tblastn.mafft.rmdup.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# phyogenetic analysis using iqtree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-03-23T10:42:53.521425Z",
     "start_time": "2024-03-23T06:59:17.747580Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IQ-TREE multicore version 2.1.4-beta COVID-edition for Mac OS X 64-bit built Jun 24 2021\n",
      "Developed by Bui Quang Minh, James Barbetti, Nguyen Lam Tung,\n",
      "Olga Chernomor, Heiko Schmidt, Dominik Schrempf, Michael Woodhams.\n",
      "\n",
      "Host:    j-macstudio.nih.go.jp (SSE4.2, 128 GB RAM)\n",
      "Command: iqtree -s output_examples/Coronaviridae/tblastn.mafft.rmdup.fasta -m TESTNEW -redo -nt AUTO\n",
      "Seed:    289110 (Using SPRNG - Scalable Parallel Random Number Generator)\n",
      "Time:    Thu May  2 17:07:45 2024\n",
      "Kernel:  SSE2 - auto-detect threads (20 CPU cores detected)\n",
      "\n",
      "Reading alignment file output_examples/Coronaviridae/tblastn.mafft.rmdup.fasta ... Fasta format detected\n",
      "Alignment most likely contains protein sequences\n",
      "WARNING: 5 sites contain only gaps or ambiguous characters.\n",
      "Alignment has 169 sequences with 316 columns, 105 distinct patterns\n",
      "19 parsimony-informative, 70 singleton sites, 227 constant sites\n",
      "          Gap/Ambiguity  Composition  p-value\n",
      "   1  BS007339    1.58%    passed    100.00%\n",
      "   2  BS007341    1.58%    passed    100.00%\n",
      "   3  BS007342    1.58%    passed    100.00%\n",
      "   4  BS007388    1.58%    passed    100.00%\n",
      "   5  BS007463    1.58%    passed    100.00%\n",
      "   6  HG994235    1.58%    passed    100.00%\n",
      "   7  LC663793    1.58%    passed    100.00%\n",
      "   8  MN908947    1.58%    passed    100.00%\n",
      "   9  MT050493    1.58%    passed    100.00%\n",
      "  10  MT072864    1.58%    passed    100.00%\n",
      "  11  MT121216    1.58%    passed    100.00%\n",
      "  12  MT635444    1.58%    passed    100.00%\n",
      "  13  MW251308    1.58%    passed    100.00%\n",
      "  14  MW273805    1.58%    passed    100.00%\n",
      "  15  MW294011    1.58%    passed    100.00%\n",
      "  16  MW356785    1.58%    passed    100.00%\n",
      "  17  MW643223    1.58%    passed    100.00%\n",
      "  18  MW681002    1.58%    passed    100.00%\n",
      "  19  MW703458    1.90%    passed    100.00%\n",
      "  20  MW719567    1.58%    passed    100.00%\n",
      "  21  MW816472    1.58%    passed    100.00%\n",
      "  22  MW975592    1.58%    passed    100.00%\n",
      "  23  MW976746    1.58%    passed    100.00%\n",
      "  24  MZ054862    1.58%    passed    100.00%\n",
      "  25  MZ081376    1.58%    passed    100.00%\n",
      "  26  MZ081377    1.58%    passed    100.00%\n",
      "  27  MZ081378    1.58%    passed    100.00%\n",
      "  28  MZ081379    1.58%    passed    100.00%\n",
      "  29  MZ081380    1.58%    passed    100.00%\n",
      "  30  MZ081381    1.58%    passed    100.00%\n",
      "  31  MZ081382    1.58%    passed    100.00%\n",
      "  32  MZ190137    1.58%    passed    100.00%\n",
      "  33  MZ190138    1.58%    passed    100.00%\n",
      "  34  MZ410620    1.58%    passed    100.00%\n",
      "  35  OK017793    1.58%    passed    100.00%\n",
      "  36  OK017794    1.58%    passed    100.00%\n",
      "  37  OK017795    1.58%    passed    100.00%\n",
      "  38  OK017796    1.58%    passed    100.00%\n",
      "  39  OK017797    1.58%    passed    100.00%\n",
      "  40  OK017798    1.58%    passed    100.00%\n",
      "  41  OK017799    1.58%    passed    100.00%\n",
      "  42  OK017800    1.58%    passed    100.00%\n",
      "  43  OK017801    1.58%    passed    100.00%\n",
      "  44  OK017802    1.58%    passed    100.00%\n",
      "  45  OK017803    1.58%    passed    100.00%\n",
      "  46  OK017804    1.58%    passed    100.00%\n",
      "  47  OK017805    1.58%    passed    100.00%\n",
      "  48  OK017806    1.58%    passed    100.00%\n",
      "  49  OK017807    1.58%    passed    100.00%\n",
      "  50  OK017808    1.58%    passed    100.00%\n",
      "  51  OK017809    1.58%    passed    100.00%\n",
      "  52  OK017810    1.58%    passed    100.00%\n",
      "  53  OK017811    1.58%    passed    100.00%\n",
      "  54  OK017825    1.58%    passed    100.00%\n",
      "  55  OK017826    1.58%    passed    100.00%\n",
      "  56  OK017827    1.58%    passed    100.00%\n",
      "  57  OK017828    1.58%    passed    100.00%\n",
      "  58  OK017829    1.58%    passed    100.00%\n",
      "  59  OK017830    1.58%    passed    100.00%\n",
      "  60  OK017831    1.58%    passed    100.00%\n",
      "  61  OK017832    1.58%    passed    100.00%\n",
      "  62  OK017833    1.58%    passed    100.00%\n",
      "  63  OK017834    1.58%    passed    100.00%\n",
      "  64  OK017835    1.58%    passed    100.00%\n",
      "  65  OK017836    1.58%    passed    100.00%\n",
      "  66  OK017837    1.58%    passed    100.00%\n",
      "  67  OK017838    1.58%    passed    100.00%\n",
      "  68  OK017839    1.58%    passed    100.00%\n",
      "  69  OK017840    1.58%    passed    100.00%\n",
      "  70  OK017841    1.58%    passed    100.00%\n",
      "  71  OK017842    1.58%    passed    100.00%\n",
      "  72  OK017843    1.58%    passed    100.00%\n",
      "  73  OK017844    1.58%    passed    100.00%\n",
      "  74  OK017845    1.58%    passed    100.00%\n",
      "  75  OK017852    1.58%    passed    100.00%\n",
      "  76  OK017853    1.58%    passed    100.00%\n",
      "  77  OK017854    1.58%    passed    100.00%\n",
      "  78  OK017855    1.58%    passed    100.00%\n",
      "  79  OK017856    1.58%    passed    100.00%\n",
      "  80  OK017857    1.58%    passed    100.00%\n",
      "  81  OK017858    1.58%    passed    100.00%\n",
      "  82  OK017859    1.58%    passed    100.00%\n",
      "  83  OK017860    1.58%    passed    100.00%\n",
      "  84  OK287354    1.58%    passed    100.00%\n",
      "  85  OK287355    1.58%    passed    100.00%\n",
      "  86  OK426405    1.58%    passed    100.00%\n",
      "  87  OL549280    1.58%    passed    100.00%\n",
      "  88  OL601954    1.58%    passed    100.00%\n",
      "  89  OM739392    1.58%    passed    100.00%\n",
      "  90  OM765403    1.58%    passed    100.00%\n",
      "  91  OM811375    1.58%    passed    100.00%\n",
      "  92  ON378802    1.58%    passed    100.00%\n",
      "  93  ON797362    1.58%    passed    100.00%\n",
      "  94  OP089057    1.58%    passed    100.00%\n",
      "  95  OP288969    1.58%    passed     99.78%\n",
      "  96  OP303169    1.58%    passed    100.00%\n",
      "  97  OP430919    1.58%    passed    100.00%\n",
      "  98  OP581529    1.58%    passed    100.00%\n",
      "  99  OP776338    1.58%    passed    100.00%\n",
      " 100  OP776339    1.58%    passed    100.00%\n",
      " 101  OP776340    1.58%    passed    100.00%\n",
      " 102  OQ318691    1.58%    passed    100.00%\n",
      " 103  OQ401247    1.58%    passed    100.00%\n",
      " 104  OQ401248    1.58%    passed    100.00%\n",
      " 105  OQ401249    1.58%    passed    100.00%\n",
      " 106  OQ503498    1.58%    passed    100.00%\n",
      " 107  OQ503499    1.58%    passed    100.00%\n",
      " 108  OQ503506    1.58%    passed    100.00%\n",
      " 109  OQ564944    1.58%    passed    100.00%\n",
      " 110  OQ587537    1.58%    passed    100.00%\n",
      " 111  OQ691560    1.58%    passed    100.00%\n",
      " 112  OQ701291    1.58%    passed    100.00%\n",
      " 113  OQ847651    1.58%    passed    100.00%\n",
      " 114  OQ921116    1.58%    passed    100.00%\n",
      " 115  OQ921514    1.58%    passed    100.00%\n",
      " 116  OQ946921    1.58%    passed    100.00%\n",
      " 117  OQ977047    1.58%    passed    100.00%\n",
      " 118  OR098915    1.58%    passed    100.00%\n",
      " 119  OR098921    4.11%    passed    100.00%\n",
      " 120  OR122618    1.58%    passed    100.00%\n",
      " 121  OR122627    1.58%    passed    100.00%\n",
      " 122  OR145069    1.58%    passed    100.00%\n",
      " 123  OR196300    1.58%    passed    100.00%\n",
      " 124  OR233291    1.58%    passed    100.00%\n",
      " 125  OR233292    1.58%    passed    100.00%\n",
      " 126  OR233293    1.58%    passed    100.00%\n",
      " 127  OR233294    1.58%    passed    100.00%\n",
      " 128  OR233296    1.58%    passed    100.00%\n",
      " 129  OR233297    1.58%    passed    100.00%\n",
      " 130  OR233298    1.58%    passed    100.00%\n",
      " 131  OR233299    1.58%    passed    100.00%\n",
      " 132  OR233300    1.58%    passed    100.00%\n",
      " 133  OR233301    1.58%    passed    100.00%\n",
      " 134  OR233302    1.58%    passed    100.00%\n",
      " 135  OR233303    1.58%    passed    100.00%\n",
      " 136  OR233304    1.58%    passed    100.00%\n",
      " 137  OR233305    1.58%    passed    100.00%\n",
      " 138  OR233308    1.58%    passed    100.00%\n",
      " 139  OR233310    1.58%    passed    100.00%\n",
      " 140  OR233311    1.58%    passed    100.00%\n",
      " 141  OR233312    1.58%    passed    100.00%\n",
      " 142  OR233313    1.58%    passed    100.00%\n",
      " 143  OR233314    1.58%    passed    100.00%\n",
      " 144  OR233316    1.58%    passed    100.00%\n",
      " 145  OR233317    1.58%    passed    100.00%\n",
      " 146  OR233318    1.58%    passed    100.00%\n",
      " 147  OR233319    1.58%    passed    100.00%\n",
      " 148  OR233320    1.58%    passed    100.00%\n",
      " 149  OR233321    1.58%    passed    100.00%\n",
      " 150  OR233322    1.58%    passed    100.00%\n",
      " 151  OR233323    1.58%    passed    100.00%\n",
      " 152  OR233324    1.58%    passed    100.00%\n",
      " 153  OR233325    1.58%    passed    100.00%\n",
      " 154  OR233328    1.58%    passed    100.00%\n",
      " 155  OR261265    1.58%    passed    100.00%\n",
      " 156  OR415190    1.58%    passed    100.00%\n",
      " 157  OR462807    1.58%    passed    100.00%\n",
      " 158  OR559085    1.58%    passed    100.00%\n",
      " 159  OR565116    1.58%    passed    100.00%\n",
      " 160  OR577043    1.58%    passed    100.00%\n",
      " 161  OR577095    1.58%    passed    100.00%\n",
      " 162  OR600070    1.58%    passed    100.00%\n",
      " 163  OR603730    1.58%    passed    100.00%\n",
      " 164  OY468386    1.58%    passed    100.00%\n",
      " 165  OY726593    1.58%    passed    100.00%\n",
      " 166  OY731253    1.58%    passed    100.00%\n",
      " 167  OY739714    1.58%    passed    100.00%\n",
      " 168  OY742622    1.58%    passed    100.00%\n",
      " 169  OY742935    1.58%    passed    100.00%\n",
      "****  TOTAL       1.60%  0 sequences failed composition chi2 test (p-value<5%; df=19)\n",
      "NOTE: BS007342 is identical to BS007339 but kept for subsequent analysis\n",
      "NOTE: MN908947 is identical to BS007341 but kept for subsequent analysis\n",
      "NOTE: OK017838 is identical to MW681002 but kept for subsequent analysis\n",
      "NOTE: OP776338 is identical to MW719567 but kept for subsequent analysis\n",
      "NOTE: MZ081378 is identical to MZ081376 but kept for subsequent analysis\n",
      "NOTE: MZ081379 is identical to MZ081377 but kept for subsequent analysis\n",
      "NOTE: OK017795 is identical to OK017794 but kept for subsequent analysis\n",
      "NOTE: OK017844 is identical to OK017808 but kept for subsequent analysis\n",
      "NOTE: OK017810 is identical to OK017809 but kept for subsequent analysis\n",
      "NOTE: OP776340 is identical to OP776339 but kept for subsequent analysis\n",
      "NOTE: OR233311 is identical to OR233301 but kept for subsequent analysis\n",
      "NOTE: 118 identical sequences (see below) will be ignored for subsequent analysis\n",
      "NOTE: BS007388 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: BS007463 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: OQ587537 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: OQ701291 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: OQ921116 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: OQ946921 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: OQ977047 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: OR098915 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: OR122618 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: OR122627 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: OR145069 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: OR196300 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: OR415190 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: OR462807 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: OR559085 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: OR565116 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: OR577043 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: OR600070 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: OR603730 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: OY468386 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: OY726593 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: OY731253 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: OY739714 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: OY742622 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: OY742935 (identical to BS007339) is ignored but added at the end\n",
      "NOTE: MT050493 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: MT635444 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: MW251308 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: MW273805 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: MW294011 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: MW356785 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: MW643223 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: MW975592 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: MW976746 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: MZ054862 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: MZ081381 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: MZ410620 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: OK426405 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: OL549280 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: OL601954 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: OM739392 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: OM765403 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: OM811375 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: ON797362 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: OP089057 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: OP303169 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: OP581529 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: OQ318691 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: OQ691560 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: OQ847651 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: OQ921514 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: OR233322 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: OR233323 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: OR233324 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: OR233328 (identical to BS007341) is ignored but added at the end\n",
      "NOTE: OK017840 (identical to MW681002) is ignored but added at the end\n",
      "NOTE: OK017841 (identical to MW681002) is ignored but added at the end\n",
      "NOTE: OK017842 (identical to MW681002) is ignored but added at the end\n",
      "NOTE: OK017843 (identical to MW681002) is ignored but added at the end\n",
      "NOTE: OK287355 (identical to MW681002) is ignored but added at the end\n",
      "NOTE: OR233298 (identical to MW681002) is ignored but added at the end\n",
      "NOTE: OR233299 (identical to MW681002) is ignored but added at the end\n",
      "NOTE: MZ081380 (identical to MZ081376) is ignored but added at the end\n",
      "NOTE: OK017793 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OK017801 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OK017802 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OK017803 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OK017804 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OK017825 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OK017826 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OK017827 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OK017828 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OK017830 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OK017831 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OK017833 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OK017836 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OK017852 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OK017853 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OK017854 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OK017855 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OK017856 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OK017857 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OK017859 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OK017860 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OQ503498 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OQ503499 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OQ503506 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OR233292 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OR233293 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OR233294 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OR233296 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OR233297 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OR233300 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OR233303 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OR233304 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OR233305 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OR233308 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OR233310 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OR233316 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OR233317 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OR233318 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OR233319 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OR233320 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OR233321 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OR233325 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OR261265 (identical to MZ081377) is ignored but added at the end\n",
      "NOTE: OK017796 (identical to OK017794) is ignored but added at the end\n",
      "NOTE: OK017797 (identical to OK017794) is ignored but added at the end\n",
      "NOTE: OK017798 (identical to OK017794) is ignored but added at the end\n",
      "NOTE: OK017799 (identical to OK017794) is ignored but added at the end\n",
      "NOTE: OK017800 (identical to OK017794) is ignored but added at the end\n",
      "NOTE: OK017811 (identical to OK017809) is ignored but added at the end\n",
      "NOTE: OQ401247 (identical to OP776339) is ignored but added at the end\n",
      "NOTE: OQ401248 (identical to OP776339) is ignored but added at the end\n",
      "NOTE: OQ401249 (identical to OP776339) is ignored but added at the end\n",
      "NOTE: OR233312 (identical to OR233301) is ignored but added at the end\n",
      "NOTE: OR233313 (identical to OR233301) is ignored but added at the end\n",
      "NOTE: OR233314 (identical to OR233301) is ignored but added at the end\n",
      "Alignment was printed to output_examples/Coronaviridae/tblastn.mafft.rmdup.fasta.uniqueseq.phy\n",
      "\n",
      "For your convenience alignment with unique sequences printed to output_examples/Coronaviridae/tblastn.mafft.rmdup.fasta.uniqueseq.phy\n",
      "\n",
      "\n",
      "Create initial parsimony tree by phylogenetic likelihood library (PLL)... 0.003 seconds\n",
      "Measuring multi-threading efficiency up to 20 CPU cores\n",
      "Increase to 10 rounds for branch lengths\n",
      "208 trees examined\n",
      "Threads: 1 / Time: 20.057 sec / Speedup: 1.000 / Efficiency: 100% / LogL: -2250\n",
      "Threads: 2 / Time: 11.976 sec / Speedup: 1.675 / Efficiency: 84% / LogL: -2250\n",
      "Threads: 3 / Time: 10.056 sec / Speedup: 1.995 / Efficiency: 66% / LogL: -2250\n",
      "Threads: 4 / Time: 9.551 sec / Speedup: 2.100 / Efficiency: 53% / LogL: -2250\n",
      "Threads: 5 / Time: 8.785 sec / Speedup: 2.283 / Efficiency: 46% / LogL: -2250\n",
      "BEST NUMBER OF THREADS: 4\n",
      "\n",
      "Perform fast likelihood tree search using LG+I+G model...\n",
      "Estimate model parameters (epsilon = 5.000)\n",
      "Perform nearest neighbor interchange...\n",
      "Estimate model parameters (epsilon = 1.000)\n",
      "1. Initial log-likelihood: -1684.078\n",
      "Optimal log-likelihood: -1684.074\n",
      "Proportion of invariable sites: 0.357\n",
      "Gamma shape alpha: 0.758\n",
      "Parameters optimization took 1 rounds (0.016 sec)\n",
      "Time for fast ML tree search: 0.102 seconds\n",
      "\n",
      "NOTE: ModelFinder requires 9 MB RAM!\n",
      "ModelFinder will test up to 702 protein models (sample size: 316) ...\n",
      " No. Model         -LnL         df  AIC          AICc         BIC\n",
      "  1  LG            1710.413     99  3618.826     3710.493     3990.645\n",
      "  2  LG+I          1688.119     100 3576.238     3670.192     3951.812\n",
      "  3  LG+G4         1684.981     100 3569.962     3663.916     3945.536\n",
      "  4  LG+I+G4       1684.074     101 3570.148     3666.429     3949.478\n",
      "  5  LG+R2         1681.316     101 3564.632     3660.913     3943.962\n",
      "  6  LG+R3         1681.145     103 3568.290     3669.347     3955.131\n",
      " 15  LG+F+I        1661.439     119 3560.878     3706.593     4007.812\n",
      " 16  LG+F+G4       1658.357     119 3554.714     3700.429     4001.648\n",
      " 17  LG+F+I+G4     1657.432     120 3554.865     3703.788     4005.554\n",
      " 18  LG+F+R2       1654.643     120 3549.285     3698.208     3999.974\n",
      " 28  WAG+I         1690.608     100 3581.216     3675.170     3956.790\n",
      " 29  WAG+G4        1687.024     100 3574.049     3668.002     3949.623\n",
      " 30  WAG+I+G4      1686.177     101 3574.353     3670.633     3953.683\n",
      " 31  WAG+R2        1682.311     101 3566.622     3662.903     3945.952\n",
      " 41  WAG+F+I       1665.050     119 3568.100     3713.815     4015.034\n",
      " 42  WAG+F+G4      1661.658     119 3561.316     3707.030     4008.249\n",
      " 43  WAG+F+I+G4    1660.824     120 3561.648     3710.572     4012.337\n",
      " 44  WAG+F+R2      1656.876     120 3553.752     3702.675     4004.441\n",
      " 54  JTT+I         1694.537     100 3589.075     3683.028     3964.649\n",
      " 55  JTT+G4        1691.040     100 3582.081     3676.034     3957.655\n",
      " 56  JTT+I+G4      1690.010     101 3582.020     3678.300     3961.350\n",
      " 57  JTT+R2        1687.265     101 3576.530     3672.810     3955.860\n",
      " 67  JTT+F+I       1669.038     119 3576.075     3721.789     4023.008\n",
      " 68  JTT+F+G4      1665.787     119 3569.574     3715.288     4016.507\n",
      " 69  JTT+F+I+G4    1664.757     120 3569.514     3718.437     4020.203\n",
      " 70  JTT+F+R2      1662.404     120 3564.809     3713.732     4015.498\n",
      " 80  Q.pfam+I      1691.025     100 3582.050     3676.003     3957.624\n",
      " 81  Q.pfam+G4     1687.889     100 3575.777     3669.730     3951.351\n",
      " 82  Q.pfam+I+G4   1687.027     101 3576.054     3672.335     3955.384\n",
      " 83  Q.pfam+R2     1683.819     101 3569.637     3665.918     3948.967\n",
      " 93  Q.pfam+F+I    1662.292     119 3562.583     3708.297     4009.516\n",
      " 94  Q.pfam+F+G4   1659.312     119 3556.623     3702.338     4003.557\n",
      " 95  Q.pfam+F+I+G4 1658.455     120 3556.911     3705.834     4007.600\n",
      " 96  Q.pfam+F+R2   1655.230     120 3550.461     3699.384     4001.150\n",
      "106  Q.bird+I      1717.246     100 3634.491     3728.445     4010.065\n",
      "107  Q.bird+G4     1713.580     100 3627.160     3721.114     4002.734\n",
      "108  Q.bird+I+G4   1712.222     101 3626.444     3722.725     4005.774\n",
      "109  Q.bird+R2     1711.398     101 3624.797     3721.077     4004.127\n",
      "119  Q.bird+F+I    1682.114     119 3602.228     3747.943     4049.162\n",
      "120  Q.bird+F+G4   1679.045     119 3596.089     3741.803     4043.023\n",
      "121  Q.bird+F+I+G4 1677.719     120 3595.439     3744.362     4046.128\n",
      "122  Q.bird+F+R2   1677.867     120 3595.735     3744.658     4046.424\n",
      "132  Q.mammal+I    1715.722     100 3631.445     3725.398     4007.019\n",
      "133  Q.mammal+G4   1712.153     100 3624.305     3718.259     3999.879\n",
      "134  Q.mammal+I+G4 1710.897     101 3623.794     3720.075     4003.124\n",
      "135  Q.mammal+R2   1709.245     101 3620.491     3716.771     3999.821\n",
      "145  Q.mammal+F+I  1675.958     119 3589.917     3735.631     4036.850\n",
      "146  Q.mammal+F+G4 1673.072     119 3584.145     3729.859     4031.078\n",
      "147  Q.mammal+F+I+G4 1671.866     120 3583.732     3732.655     4034.421\n",
      "148  Q.mammal+F+R2 1671.327     120 3582.655     3731.578     4033.344\n",
      "158  Q.insect+I    1689.818     100 3579.636     3673.589     3955.210\n",
      "159  Q.insect+G4   1686.738     100 3573.476     3667.430     3949.050\n",
      "160  Q.insect+I+G4 1685.668     101 3573.335     3669.616     3952.665\n",
      "161  Q.insect+R2   1683.076     101 3568.152     3664.432     3947.482\n",
      "171  Q.insect+F+I  1663.882     119 3565.764     3711.479     4012.698\n",
      "172  Q.insect+F+G4 1660.922     119 3559.845     3705.559     4006.778\n",
      "173  Q.insect+F+I+G4 1659.850     120 3559.700     3708.623     4010.389\n",
      "174  Q.insect+F+R2 1657.461     120 3554.922     3703.845     4005.611\n",
      "184  Q.plant+I     1696.678     100 3593.355     3687.309     3968.929\n",
      "185  Q.plant+G4    1693.045     100 3586.091     3680.044     3961.665\n",
      "186  Q.plant+I+G4  1691.717     101 3585.434     3681.714     3964.764\n",
      "187  Q.plant+R2    1690.048     101 3582.096     3678.377     3961.426\n",
      "197  Q.plant+F+I   1667.004     119 3572.007     3717.722     4018.941\n",
      "198  Q.plant+F+G4  1663.653     119 3565.305     3711.020     4012.239\n",
      "199  Q.plant+F+I+G4 1662.288     120 3564.576     3713.499     4015.265\n",
      "200  Q.plant+F+R2  1660.837     120 3561.675     3710.598     4012.364\n",
      "210  Q.yeast+I     1695.894     100 3591.787     3685.741     3967.361\n",
      "211  Q.yeast+G4    1692.819     100 3585.639     3679.592     3961.213\n",
      "212  Q.yeast+I+G4  1691.762     101 3585.525     3681.805     3964.855\n",
      "213  Q.yeast+R2    1689.269     101 3580.537     3676.818     3959.867\n",
      "223  Q.yeast+F+I   1664.147     119 3566.293     3712.007     4013.226\n",
      "224  Q.yeast+F+G4  1661.230     119 3560.461     3706.175     4007.394\n",
      "225  Q.yeast+F+I+G4 1660.215     120 3560.430     3709.353     4011.119\n",
      "226  Q.yeast+F+R2  1657.868     120 3555.736     3704.659     4006.425\n",
      "236  JTTDCMut+I    1694.389     100 3588.778     3682.731     3964.352\n",
      "237  JTTDCMut+G4   1690.860     100 3581.720     3675.674     3957.294\n",
      "238  JTTDCMut+I+G4 1689.814     101 3581.629     3677.909     3960.958\n",
      "239  JTTDCMut+R2   1687.115     101 3576.230     3672.511     3955.560\n",
      "249  JTTDCMut+F+I  1668.989     119 3575.979     3721.693     4022.912\n",
      "250  JTTDCMut+F+G4 1665.716     119 3569.432     3715.146     4016.365\n",
      "251  JTTDCMut+F+I+G4 1664.667     120 3569.334     3718.257     4020.023\n",
      "252  JTTDCMut+F+R2 1662.379     120 3564.759     3713.682     4015.448\n",
      "262  DCMut+I       1712.839     100 3625.678     3719.631     4001.252\n",
      "263  DCMut+G4      1707.588     100 3615.177     3709.130     3990.751\n",
      "264  DCMut+I+G4    1706.125     101 3614.250     3710.530     3993.580\n",
      "265  DCMut+R2      1702.897     101 3607.794     3704.074     3987.124\n",
      "275  DCMut+F+I     1679.177     119 3596.354     3742.068     4043.287\n",
      "276  DCMut+F+G4    1674.482     119 3586.965     3732.679     4033.898\n",
      "277  DCMut+F+I+G4  1673.035     120 3586.070     3734.993     4036.759\n",
      "278  DCMut+F+R2    1670.119     120 3580.238     3729.161     4030.927\n",
      "288  VT+I          1697.069     100 3594.139     3688.092     3969.713\n",
      "289  VT+G4         1693.832     100 3587.663     3681.617     3963.237\n",
      "290  VT+I+G4       1693.101     101 3588.203     3684.483     3967.533\n",
      "291  VT+R2         1688.642     101 3579.285     3675.565     3958.615\n",
      "301  VT+F+I        1671.600     119 3581.199     3726.913     4028.132\n",
      "302  VT+F+G4       1668.423     119 3574.845     3720.560     4021.779\n",
      "303  VT+F+I+G4     1667.704     120 3575.409     3724.332     4026.098\n",
      "304  VT+F+R2       1663.059     120 3566.118     3715.041     4016.807\n",
      "314  PMB+I         1701.129     100 3602.258     3696.212     3977.832\n",
      "315  PMB+G4        1697.793     100 3595.586     3689.539     3971.160\n",
      "316  PMB+I+G4      1697.059     101 3596.119     3692.399     3975.449\n",
      "317  PMB+R2        1691.388     101 3584.775     3681.056     3964.105\n",
      "327  PMB+F+I       1678.996     119 3595.993     3741.707     4042.926\n",
      "328  PMB+F+G4      1675.629     119 3589.258     3734.972     4036.191\n",
      "329  PMB+F+I+G4    1674.895     120 3589.790     3738.713     4040.479\n",
      "330  PMB+F+R2      1668.771     120 3577.543     3726.466     4028.232\n",
      "340  Blosum62+I    1697.577     100 3595.154     3689.108     3970.729\n",
      "341  Blosum62+G4   1694.298     100 3588.596     3682.549     3964.170\n",
      "342  Blosum62+I+G4 1693.547     101 3589.094     3685.375     3968.424\n",
      "343  Blosum62+R2   1688.429     101 3578.858     3675.138     3958.188\n",
      "353  Blosum62+F+I  1679.045     119 3596.090     3741.804     4043.023\n",
      "354  Blosum62+F+G4 1675.778     119 3589.556     3735.270     4036.489\n",
      "355  Blosum62+F+I+G4 1675.034     120 3590.069     3738.992     4040.758\n",
      "356  Blosum62+F+R2 1669.620     120 3579.239     3728.162     4029.928\n",
      "366  Dayhoff+I     1712.687     100 3625.374     3719.327     4000.948\n",
      "367  Dayhoff+G4    1707.400     100 3614.799     3708.753     3990.374\n",
      "368  Dayhoff+I+G4  1705.931     101 3613.863     3710.143     3993.193\n",
      "369  Dayhoff+R2    1702.676     101 3607.352     3703.633     3986.682\n",
      "379  Dayhoff+F+I   1679.047     119 3596.095     3741.809     4043.028\n",
      "380  Dayhoff+F+G4  1674.321     119 3586.641     3732.355     4033.574\n",
      "381  Dayhoff+F+I+G4 1672.868     120 3585.737     3734.660     4036.426\n",
      "382  Dayhoff+F+R2  1669.923     120 3579.847     3728.770     4030.536\n",
      "392  mtREV+I       1763.012     100 3726.025     3819.978     4101.599\n",
      "393  mtREV+G4      1758.843     100 3717.687     3811.640     4093.261\n",
      "394  mtREV+I+G4    1757.643     101 3717.286     3813.566     4096.616\n",
      "395  mtREV+R2      1757.747     101 3717.493     3813.774     4096.823\n",
      "405  mtREV+F+I     1680.244     119 3598.487     3744.201     4045.421\n",
      "406  mtREV+F+G4    1675.614     119 3589.229     3734.943     4036.162\n",
      "407  mtREV+F+I+G4  1674.305     120 3588.610     3737.533     4039.299\n",
      "408  mtREV+F+R2    1673.478     120 3586.955     3735.878     4037.644\n",
      "418  mtART+I       1762.188     100 3724.375     3818.329     4099.949\n",
      "419  mtART+G4      1756.313     100 3712.625     3806.579     4088.199\n",
      "420  mtART+I+G4    1753.129     101 3708.259     3804.539     4087.589\n",
      "421  mtART+R2      1752.323     101 3706.647     3802.927     4085.977\n",
      "431  mtART+F+I     1682.282     119 3602.565     3748.279     4049.498\n",
      "432  mtART+F+G4    1677.255     119 3592.510     3738.224     4039.443\n",
      "433  mtART+F+I+G4  1675.096     120 3590.191     3739.115     4040.881\n",
      "434  mtART+F+R2    1674.186     120 3588.372     3737.295     4039.061\n",
      "444  mtZOA+I       1745.993     100 3691.985     3785.939     4067.560\n",
      "445  mtZOA+G4      1741.385     100 3682.771     3776.724     4058.345\n",
      "446  mtZOA+I+G4    1739.107     101 3680.215     3776.495     4059.545\n",
      "447  mtZOA+R2      1739.041     101 3680.081     3776.362     4059.411\n",
      "457  mtZOA+F+I     1670.781     119 3579.561     3725.276     4026.495\n",
      "458  mtZOA+F+G4    1666.539     119 3571.078     3716.792     4018.011\n",
      "459  mtZOA+F+I+G4  1665.001     120 3570.002     3718.925     4020.691\n",
      "460  mtZOA+F+R2    1664.057     120 3568.115     3717.038     4018.804\n",
      "470  mtMet+I       1742.692     100 3685.383     3779.337     4060.958\n",
      "471  mtMet+G4      1739.160     100 3678.321     3772.274     4053.895\n",
      "472  mtMet+I+G4    1737.633     101 3677.265     3773.546     4056.595\n",
      "473  mtMet+R2      1737.985     101 3677.970     3774.251     4057.300\n",
      "483  mtMet+F+I     1672.860     119 3583.720     3729.434     4030.653\n",
      "484  mtMet+F+G4    1669.516     119 3577.031     3722.745     4023.965\n",
      "485  mtMet+F+I+G4  1668.270     120 3576.540     3725.463     4027.229\n",
      "486  mtMet+F+R2    1667.859     120 3575.717     3724.641     4026.407\n",
      "496  mtVer+I       1772.551     100 3745.102     3839.055     4120.676\n",
      "497  mtVer+G4      1769.011     100 3738.021     3831.975     4113.595\n",
      "498  mtVer+I+G4    1767.444     101 3736.887     3833.168     4116.217\n",
      "499  mtVer+R2      1768.811     101 3739.622     3835.902     4118.952\n",
      "509  mtVer+F+I     1685.769     119 3609.538     3755.253     4056.472\n",
      "510  mtVer+F+G4    1681.853     119 3601.707     3747.421     4048.640\n",
      "511  mtVer+F+I+G4  1680.268     120 3600.536     3749.459     4051.225\n",
      "512  mtVer+F+R2    1680.405     120 3600.809     3749.732     4051.498\n",
      "522  mtInv+I       1751.945     100 3703.891     3797.844     4079.465\n",
      "523  mtInv+G4      1748.494     100 3696.988     3790.942     4072.563\n",
      "524  mtInv+I+G4    1747.062     101 3696.124     3792.404     4075.454\n",
      "525  mtInv+R2      1745.699     101 3693.399     3789.679     4072.729\n",
      "535  mtInv+F+I     1669.494     119 3576.988     3722.702     4023.921\n",
      "536  mtInv+F+G4    1666.049     119 3570.097     3715.811     4017.030\n",
      "537  mtInv+F+I+G4  1664.884     120 3569.768     3718.691     4020.457\n",
      "538  mtInv+F+R2    1663.197     120 3566.393     3715.316     4017.082\n",
      "548  mtMAM+I       1777.914     100 3755.828     3849.781     4131.402\n",
      "549  mtMAM+G4      1774.410     100 3748.819     3842.773     4124.393\n",
      "550  mtMAM+I+G4    1772.201     101 3746.402     3842.682     4125.732\n",
      "551  mtMAM+R2      1776.757     101 3755.515     3851.795     4134.845\n",
      "561  mtMAM+F+I     1696.495     119 3630.990     3776.704     4077.924\n",
      "562  mtMAM+F+G4    1692.599     119 3623.198     3768.912     4070.131\n",
      "563  mtMAM+F+I+G4  1690.685     120 3621.370     3770.293     4072.059\n",
      "564  mtMAM+F+R2    1693.683     120 3627.365     3776.288     4078.054\n",
      "574  HIVb+I        1714.982     100 3629.964     3723.917     4005.538\n",
      "575  HIVb+G4       1710.502     100 3621.005     3714.958     3996.579\n",
      "576  HIVb+I+G4     1709.016     101 3620.032     3716.312     3999.362\n",
      "577  HIVb+R2       1707.691     101 3617.381     3713.662     3996.711\n",
      "587  HIVb+F+I      1674.045     119 3586.091     3731.805     4033.024\n",
      "588  HIVb+F+G4     1669.865     119 3577.731     3723.445     4024.664\n",
      "589  HIVb+F+I+G4   1668.300     120 3576.600     3725.523     4027.289\n",
      "590  HIVb+F+R2     1667.887     120 3575.773     3724.696     4026.462\n",
      "600  HIVw+I        1753.912     100 3707.824     3801.778     4083.399\n",
      "601  HIVw+G4       1747.302     100 3694.604     3788.557     4070.178\n",
      "602  HIVw+I+G4     1744.324     101 3690.648     3786.928     4069.978\n",
      "603  HIVw+R2       1743.018     101 3688.036     3784.317     4067.366\n",
      "613  HIVw+F+I      1701.422     119 3640.845     3786.559     4087.778\n",
      "614  HIVw+F+G4     1696.969     119 3631.939     3777.653     4078.872\n",
      "615  HIVw+F+I+G4   1694.524     120 3629.048     3777.971     4079.737\n",
      "616  HIVw+F+R2     1694.541     120 3629.082     3778.005     4079.771\n",
      "626  FLU+I         1711.417     100 3622.833     3716.787     3998.408\n",
      "627  FLU+G4        1706.762     100 3613.525     3707.478     3989.099\n",
      "628  FLU+I+G4      1704.940     101 3611.880     3708.160     3991.210\n",
      "629  FLU+R2        1703.469     101 3608.939     3705.219     3988.269\n",
      "639  FLU+F+I       1684.924     119 3607.848     3753.563     4054.782\n",
      "640  FLU+F+G4      1680.980     119 3599.961     3745.675     4046.894\n",
      "641  FLU+F+I+G4    1679.273     120 3598.546     3747.469     4049.235\n",
      "642  FLU+F+R2      1678.551     120 3597.103     3746.026     4047.792\n",
      "652  rtREV+I       1718.287     100 3636.573     3730.527     4012.148\n",
      "653  rtREV+G4      1715.013     100 3630.027     3723.980     4005.601\n",
      "654  rtREV+I+G4    1714.168     101 3630.336     3726.616     4009.666\n",
      "655  rtREV+R2      1711.097     101 3624.194     3720.474     4003.524\n",
      "665  rtREV+F+I     1669.365     119 3576.730     3722.444     4023.663\n",
      "666  rtREV+F+G4    1666.063     119 3570.127     3715.841     4017.060\n",
      "667  rtREV+F+I+G4  1665.070     120 3570.140     3719.063     4020.829\n",
      "668  rtREV+F+R2    1661.780     120 3563.561     3712.484     4014.250\n",
      "678  cpREV+I       1706.714     100 3613.427     3707.381     3989.002\n",
      "679  cpREV+G4      1702.816     100 3605.632     3699.586     3981.207\n",
      "680  cpREV+I+G4    1701.731     101 3605.462     3701.743     3984.792\n",
      "681  cpREV+R2      1698.086     101 3598.172     3694.452     3977.502\n",
      "691  cpREV+F+I     1671.643     119 3581.286     3727.000     4028.219\n",
      "692  cpREV+F+G4    1667.351     119 3572.702     3718.416     4019.635\n",
      "693  cpREV+F+I+G4  1666.172     120 3572.344     3721.267     4023.033\n",
      "694  cpREV+F+R2    1662.299     120 3564.598     3713.521     4015.287\n",
      "Akaike Information Criterion:           LG+F+R2\n",
      "Corrected Akaike Information Criterion: LG+R2\n",
      "Bayesian Information Criterion:         LG+R2\n",
      "Best-fit model: LG+R2 chosen according to BIC\n",
      "\n",
      "All model information printed to output_examples/Coronaviridae/tblastn.mafft.rmdup.fasta.model.gz\n",
      "CPU time for ModelFinder: 192.824 seconds (0h:3m:12s)\n",
      "Wall-clock time for ModelFinder: 70.333 seconds (0h:1m:10s)\n",
      "\n",
      "NOTE: 1 MB RAM (0 GB) is required!\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "1. Initial log-likelihood: -1681.316\n",
      "Optimal log-likelihood: -1681.253\n",
      "Site proportion and rates:  (0.904,0.420) (0.096,6.466)\n",
      "Parameters optimization took 1 rounds (0.018 sec)\n",
      "Computing ML distances based on estimated model parameters...\n",
      "Computing ML distances took 0.015747 sec (of wall-clock time) 0.059645 sec(of CPU time)\n",
      "Computing RapidNJ tree took 0.002867 sec (of wall-clock time) 0.008555 sec (of CPU time)\n",
      "Log-likelihood of RapidNJ tree: -1679.430\n",
      "--------------------------------------------------------------------\n",
      "|             INITIALIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Generating 98 parsimony trees... 0.178 second\n",
      "Computing log-likelihood of 98 initial trees ... 0.305 seconds\n",
      "Current best score: -1669.291\n",
      "\n",
      "Do NNI search on 20 best initial trees\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 1: -1669.242\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 4: -1668.453\n",
      "Iteration 10 / LogL: -1671.335 / Time: 0h:0m:1s\n",
      "Iteration 20 / LogL: -1671.917 / Time: 0h:0m:1s\n",
      "Finish initializing candidate tree set (20)\n",
      "Current best tree score: -1668.453 / CPU time: 1.774\n",
      "Number of iterations: 20\n",
      "--------------------------------------------------------------------\n",
      "|               OPTIMIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 23: -1668.081\n",
      "Iteration 30 / LogL: -1671.849 / Time: 0h:0m:2s (0h:0m:8s left)\n",
      "Iteration 40 / LogL: -1668.086 / Time: 0h:0m:3s (0h:0m:8s left)\n",
      "Iteration 50 / LogL: -1675.102 / Time: 0h:0m:4s (0h:0m:6s left)\n",
      "Iteration 60 / LogL: -1668.083 / Time: 0h:0m:5s (0h:0m:5s left)\n",
      "Iteration 70 / LogL: -1668.610 / Time: 0h:0m:6s (0h:0m:4s left)\n",
      "Iteration 80 / LogL: -1668.083 / Time: 0h:0m:7s (0h:0m:3s left)\n",
      "Iteration 90 / LogL: -1668.082 / Time: 0h:0m:8s (0h:0m:3s left)\n",
      "Iteration 100 / LogL: -1668.085 / Time: 0h:0m:9s (0h:0m:2s left)\n",
      "Iteration 110 / LogL: -1668.083 / Time: 0h:0m:10s (0h:0m:1s left)\n",
      "Iteration 120 / LogL: -1671.806 / Time: 0h:0m:10s (0h:0m:0s left)\n",
      "TREE SEARCH COMPLETED AFTER 124 ITERATIONS / Time: 0h:0m:11s\n",
      "\n",
      "--------------------------------------------------------------------\n",
      "|                    FINALIZING TREE SEARCH                        |\n",
      "--------------------------------------------------------------------\n",
      "Performs final model parameters optimization\n",
      "Estimate model parameters (epsilon = 0.010)\n",
      "1. Initial log-likelihood: -1668.081\n",
      "Optimal log-likelihood: -1668.073\n",
      "Site proportion and rates:  (0.912,0.423) (0.088,7.008)\n",
      "Parameters optimization took 1 rounds (0.016 sec)\n",
      "BEST SCORE FOUND : -1668.073\n",
      "Total tree length: 0.593\n",
      "\n",
      "Total number of iterations: 124\n",
      "CPU time used for tree search: 43.871 sec (0h:0m:43s)\n",
      "Wall-clock time used for tree search: 11.210 sec (0h:0m:11s)\n",
      "Total CPU time used: 44.204 sec (0h:0m:44s)\n",
      "Total wall-clock time used: 11.293 sec (0h:0m:11s)\n",
      "\n",
      "Analysis results written to: \n",
      "  IQ-TREE report:                output_examples/Coronaviridae/tblastn.mafft.rmdup.fasta.iqtree\n",
      "  Maximum-likelihood tree:       output_examples/Coronaviridae/tblastn.mafft.rmdup.fasta.treefile\n",
      "  Likelihood distances:          output_examples/Coronaviridae/tblastn.mafft.rmdup.fasta.mldist\n",
      "  Screen log file:               output_examples/Coronaviridae/tblastn.mafft.rmdup.fasta.log\n",
      "\n",
      "Date and Time: Thu May  2 17:09:06 2024\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.\n",
      "OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.\n",
      "OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "train_virus=\"Coronaviridae\"\n",
    "\n",
    "iqtree \\\n",
    "    -s output_examples/${train_virus}/tblastn.mafft.rmdup.fasta \\\n",
    "    -m TESTNEW \\\n",
    "    -redo \\\n",
    "    -nt AUTO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env_210819",
   "language": "python",
   "name": "env_210819"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "187.391px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
