{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Inputs\" data-toc-modified-id=\"Inputs-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Inputs</a></span></li><li><span><a href=\"#splitting-viral-genomic-sequences-into-250-bp-fragments\" data-toc-modified-id=\"splitting-viral-genomic-sequences-into-250-bp-fragments-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>splitting viral genomic sequences into 250-bp fragments</a></span></li><li><span><a href=\"#read-curated-data\" data-toc-modified-id=\"read-curated-data-3\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>read curated data</a></span></li><li><span><a href=\"#データセットごとに配列を分割\" data-toc-modified-id=\"データセットごとに配列を分割-4\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>データセットごとに配列を分割</a></span><ul class=\"toc-item\"><li><span><a href=\"#5-cross-fold-validation-datasetの作成\" data-toc-modified-id=\"5-cross-fold-validation-datasetの作成-4.1\"><span class=\"toc-item-num\">4.1&nbsp;&nbsp;</span>5-cross fold validation datasetの作成</a></span></li><li><span><a href=\"#データセット分割の可視化\" data-toc-modified-id=\"データセット分割の可視化-4.2\"><span class=\"toc-item-num\">4.2&nbsp;&nbsp;</span>データセット分割の可視化</a></span></li><li><span><a href=\"#フォルダ作成\" data-toc-modified-id=\"フォルダ作成-4.3\"><span class=\"toc-item-num\">4.3&nbsp;&nbsp;</span>フォルダ作成</a></span></li></ul></li><li><span><a href=\"#データセットごとのリード配列の数\" data-toc-modified-id=\"データセットごとのリード配列の数-5\"><span class=\"toc-item-num\">5&nbsp;&nbsp;</span>データセットごとのリード配列の数</a></span></li><li><span><a href=\"#Main\" data-toc-modified-id=\"Main-6\"><span class=\"toc-item-num\">6&nbsp;&nbsp;</span>Main</a></span><ul class=\"toc-item\"><li><span><a href=\"#配列データ読み込み\" data-toc-modified-id=\"配列データ読み込み-6.1\"><span class=\"toc-item-num\">6.1&nbsp;&nbsp;</span>配列データ読み込み</a></span></li><li><span><a href=\"#tokenizer\" data-toc-modified-id=\"tokenizer-6.2\"><span class=\"toc-item-num\">6.2&nbsp;&nbsp;</span>tokenizer</a></span></li><li><span><a href=\"#宿主ラベルの追加\" data-toc-modified-id=\"宿主ラベルの追加-6.3\"><span class=\"toc-item-num\">6.3&nbsp;&nbsp;</span>宿主ラベルの追加</a></span></li><li><span><a href=\"#既知・未知ウイルスの分割\" data-toc-modified-id=\"既知・未知ウイルスの分割-6.4\"><span class=\"toc-item-num\">6.4&nbsp;&nbsp;</span>既知・未知ウイルスの分割</a></span><ul class=\"toc-item\"><li><span><a href=\"#2018年より前後に分けて、ヒトウイルス-or-動物ウイルスが20配列以上なら、未知ウイルスを取りおく\" data-toc-modified-id=\"2018年より前後に分けて、ヒトウイルス-or-動物ウイルスが20配列以上なら、未知ウイルスを取りおく-6.4.1\"><span class=\"toc-item-num\">6.4.1&nbsp;&nbsp;</span>2018年より前後に分けて、ヒトウイルス or 動物ウイルスが20配列以上なら、未知ウイルスを取りおく</a></span></li><li><span><a href=\"#Orthomyxoviridae：未知ウイルスはマニュアルで集めないと厳しい可能性？\" data-toc-modified-id=\"Orthomyxoviridae：未知ウイルスはマニュアルで集めないと厳しい可能性？-6.4.2\"><span class=\"toc-item-num\">6.4.2&nbsp;&nbsp;</span>Orthomyxoviridae：未知ウイルスはマニュアルで集めないと厳しい可能性？</a></span></li></ul></li><li><span><a href=\"#【確認】Main\" data-toc-modified-id=\"【確認】Main-6.5\"><span class=\"toc-item-num\">6.5&nbsp;&nbsp;</span>【確認】Main</a></span></li></ul></li></ul></div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-10T11:04:02.505866Z",
     "start_time": "2024-04-10T11:04:00.776999Z"
    }
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/env python\n",
    "import sys, re\n",
    "argvs = sys.argv\n",
    "sys.setrecursionlimit(10000000)\n",
    "\n",
    "import os\n",
    "import collections\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import array\n",
    "pd.set_option(\"display.max_colwidth\", 200)\n",
    "pd.set_option(\"display.max_columns\", 200)\n",
    "pd.set_option(\"display.max_rows\", 300)\n",
    "\n",
    "import glob\n",
    "\n",
    "## visualize\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "sns.set()\n",
    "\n",
    "plt.rcParams['font.family'] = \"Arial\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-10T11:04:02.509639Z",
     "start_time": "2024-04-10T11:04:02.507799Z"
    }
   },
   "outputs": [],
   "source": [
    "virus_family = \"Coronaviridae\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# splitting viral genomic sequences into 250-bp fragments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-10T11:04:32.011964Z",
     "start_time": "2024-04-10T11:04:23.995426Z"
    },
    "run_control": {
     "marked": false
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[INFO]\u001b[0m flag -g (--remove-gaps) is switched on when using -m (--min-len) or -M (--max-len)\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "virus_family=\"Coronaviridae\"\n",
    "\n",
    "## quality control\n",
    "cat input_examples/${virus_family}.curated.fasta | \\\n",
    "    seqkit replace -s -p \"[^atcguATCGU]\" -r \"-\" \\\n",
    "    > input_examples/${virus_family}.curated.QC.fasta\n",
    "\n",
    "## split into subsequences\n",
    "cat input_examples/${virus_family}.curated.QC.fasta | \\\n",
    "    seqkit sliding -s 125 -W 250 --greedy | \\\n",
    "    seqkit seq -m 100 \\\n",
    "    > output_examples/${virus_family}.slide.250bp.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# read curated data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-10T11:07:16.187698Z",
     "start_time": "2024-04-10T11:07:16.097431Z"
    },
    "run_control": {
     "marked": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Organism_Name</th>\n",
       "      <th>SRA_Accession</th>\n",
       "      <th>Submitters</th>\n",
       "      <th>Organization</th>\n",
       "      <th>Org_location</th>\n",
       "      <th>Release_Date</th>\n",
       "      <th>Isolate</th>\n",
       "      <th>Species</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Family</th>\n",
       "      <th>Molecule_type</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence_Type</th>\n",
       "      <th>Nuc_Completeness</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Segment</th>\n",
       "      <th>Publications</th>\n",
       "      <th>Geo_Location</th>\n",
       "      <th>Country</th>\n",
       "      <th>USA</th>\n",
       "      <th>Host</th>\n",
       "      <th>Isolation_Source</th>\n",
       "      <th>Collection_Date</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>GenBank_Title</th>\n",
       "      <th>host_label</th>\n",
       "      <th>virus_ID</th>\n",
       "      <th>collection_year</th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NC_054003</th>\n",
       "      <td>Alphacoronavirus sp.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>O'Dea,M., Prada,D., Jackson,B., Boyd,V., Baker,M.</td>\n",
       "      <td>National Center for Biotechnology Information, NIH</td>\n",
       "      <td>USA</td>\n",
       "      <td>2023-05-04</td>\n",
       "      <td>WA1087</td>\n",
       "      <td>Alphacoronavirus sp.</td>\n",
       "      <td>Alphacoronavirus</td>\n",
       "      <td>Coronaviridae</td>\n",
       "      <td>ssRNA(+)</td>\n",
       "      <td>28170</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Australia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PRJNA485481</td>\n",
       "      <td>Alphacoronavirus sp. isolate WA1087, complete genome</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_054003</td>\n",
       "      <td>2018.0</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_054004</th>\n",
       "      <td>Hipposideros pomona bat coronavirus CHB25</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Li,B., Si,H.R., Zhu,Y., Yang,X.L., Anderson,D.E., Shi,Z.L., Wang,L.F., Zhou,P.</td>\n",
       "      <td>National Center for Biotechnology Information, NIH</td>\n",
       "      <td>USA</td>\n",
       "      <td>2023-05-04</td>\n",
       "      <td>CHB0025</td>\n",
       "      <td>Alphacoronavirus CHB25</td>\n",
       "      <td>Alphacoronavirus</td>\n",
       "      <td>Coronaviridae</td>\n",
       "      <td>ssRNA(+)</td>\n",
       "      <td>28169</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>China</td>\n",
       "      <td>China</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hipposideros larvatus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PRJNA485481</td>\n",
       "      <td>Hipposideros pomona bat coronavirus CHB25 isolate CHB0025, complete genome</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_054004</td>\n",
       "      <td>2018.0</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_054015</th>\n",
       "      <td>Tylonycteris bat coronavirus HKU33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lau,S.K.P., Wong,A.C.P., Zhang,L., Luk,H.K.H., Kwok,J.S.L., Ahmed,S.S., Cai,J.P., Zhao,P.S.H., Teng,J.L.L., Tsui,S.K.W., Yuen,K.Y., Woo,P.C.Y., Li,K.S.M., Cai,J.-P.</td>\n",
       "      <td>National Center for Biotechnology Information, NIH</td>\n",
       "      <td>USA</td>\n",
       "      <td>2023-05-04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Alphacoronavirus HKU33</td>\n",
       "      <td>Alphacoronavirus</td>\n",
       "      <td>Coronaviridae</td>\n",
       "      <td>ssRNA(+)</td>\n",
       "      <td>27636</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>China</td>\n",
       "      <td>China</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Tylonycteris robustula</td>\n",
       "      <td>feces</td>\n",
       "      <td>2015-09-09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PRJNA485481</td>\n",
       "      <td>Tylonycteris bat coronavirus HKU33 strain GZ151867, complete genome</td>\n",
       "      <td>0</td>\n",
       "      <td>NC_054015</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>known</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       Organism_Name SRA_Accession  \\\n",
       "Accession                                                            \n",
       "NC_054003                       Alphacoronavirus sp.           NaN   \n",
       "NC_054004  Hipposideros pomona bat coronavirus CHB25           NaN   \n",
       "NC_054015         Tylonycteris bat coronavirus HKU33           NaN   \n",
       "\n",
       "                                                                                                                                                                     Submitters  \\\n",
       "Accession                                                                                                                                                                         \n",
       "NC_054003                                                                                                                     O'Dea,M., Prada,D., Jackson,B., Boyd,V., Baker,M.   \n",
       "NC_054004                                                                                        Li,B., Si,H.R., Zhu,Y., Yang,X.L., Anderson,D.E., Shi,Z.L., Wang,L.F., Zhou,P.   \n",
       "NC_054015  Lau,S.K.P., Wong,A.C.P., Zhang,L., Luk,H.K.H., Kwok,J.S.L., Ahmed,S.S., Cai,J.P., Zhao,P.S.H., Teng,J.L.L., Tsui,S.K.W., Yuen,K.Y., Woo,P.C.Y., Li,K.S.M., Cai,J.-P.   \n",
       "\n",
       "                                                 Organization Org_location  \\\n",
       "Accession                                                                    \n",
       "NC_054003  National Center for Biotechnology Information, NIH          USA   \n",
       "NC_054004  National Center for Biotechnology Information, NIH          USA   \n",
       "NC_054015  National Center for Biotechnology Information, NIH          USA   \n",
       "\n",
       "          Release_Date  Isolate                 Species             Genus  \\\n",
       "Accession                                                                   \n",
       "NC_054003   2023-05-04   WA1087    Alphacoronavirus sp.  Alphacoronavirus   \n",
       "NC_054004   2023-05-04  CHB0025  Alphacoronavirus CHB25  Alphacoronavirus   \n",
       "NC_054015   2023-05-04      NaN  Alphacoronavirus HKU33  Alphacoronavirus   \n",
       "\n",
       "                  Family Molecule_type  Length Sequence_Type Nuc_Completeness  \\\n",
       "Accession                                                                       \n",
       "NC_054003  Coronaviridae      ssRNA(+)   28170        RefSeq         complete   \n",
       "NC_054004  Coronaviridae      ssRNA(+)   28169        RefSeq         complete   \n",
       "NC_054015  Coronaviridae      ssRNA(+)   27636        RefSeq         complete   \n",
       "\n",
       "          Genotype  Segment  Publications Geo_Location    Country  USA  \\\n",
       "Accession                                                                \n",
       "NC_054003      NaN      NaN           NaN    Australia  Australia  NaN   \n",
       "NC_054004      NaN      NaN           1.0        China      China  NaN   \n",
       "NC_054015      NaN      NaN           1.0        China      China  NaN   \n",
       "\n",
       "                             Host Isolation_Source Collection_Date BioSample  \\\n",
       "Accession                                                                      \n",
       "NC_054003                     NaN              NaN            2018       NaN   \n",
       "NC_054004   Hipposideros larvatus              NaN         2018-03       NaN   \n",
       "NC_054015  Tylonycteris robustula            feces      2015-09-09       NaN   \n",
       "\n",
       "            BioProject  \\\n",
       "Accession                \n",
       "NC_054003  PRJNA485481   \n",
       "NC_054004  PRJNA485481   \n",
       "NC_054015  PRJNA485481   \n",
       "\n",
       "                                                                        GenBank_Title  \\\n",
       "Accession                                                                               \n",
       "NC_054003                        Alphacoronavirus sp. isolate WA1087, complete genome   \n",
       "NC_054004  Hipposideros pomona bat coronavirus CHB25 isolate CHB0025, complete genome   \n",
       "NC_054015         Tylonycteris bat coronavirus HKU33 strain GZ151867, complete genome   \n",
       "\n",
       "           host_label   virus_ID  collection_year  dataset  \n",
       "Accession                                                   \n",
       "NC_054003           0  NC_054003           2018.0  unknown  \n",
       "NC_054004           0  NC_054004           2018.0  unknown  \n",
       "NC_054015           0  NC_054015           2015.0    known  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset_f = \"/\".join(map(str, [\"input_examples\", virus_family + \".curated.csv\"]))\n",
    "\n",
    "df_dataset = pd.read_csv(dataset_f, sep=\"\\t\", index_col=0)\n",
    "df_dataset.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# data splitting for five-fold stratified cross-validation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5-cross fold validation datasetの作成"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-09T02:15:37.041693Z",
     "start_time": "2023-11-09T02:15:37.029632Z"
    }
   },
   "outputs": [],
   "source": [
    "def five_cross_fold_val(df_dataset, virus):\n",
    "    from sklearn.model_selection import StratifiedGroupKFold, StratifiedKFold\n",
    "    from operator import itemgetter\n",
    "    from sklearn.model_selection import train_test_split\n",
    "\n",
    "#     df_dataset = df_dataset[df_dataset[\"Family\"] == virus]\n",
    "#     df_dataset = df_dataset.sort_values([\"Genus\", \"host_label\"])\n",
    "#     df_dataset.reset_index(drop=True, inplace=True)\n",
    "    \n",
    "#     # virus_IDごとにデータセットに分ける\n",
    "#     if \"virus_ID\" not in df_dataset.columns.tolist():\n",
    "#         df_dataset[\"virus_ID\"] = df_dataset[\"Accession\"].copy()\n",
    "    \n",
    "    # list of virus_ID\n",
    "    df_virus = df_dataset.loc[:, [\"virus_ID\", \"host_label\", \"Genus\"]].drop_duplicates()\n",
    "    df_virus.reset_index(drop=True, inplace=True)\n",
    "    X = df_virus[\"virus_ID\"].values.tolist()\n",
    "\n",
    "    ## define class based on the combination of infectivity label and virus genus\n",
    "    df_genus = df_dataset.loc[:, [\"host_label\", \"Genus\"]].drop_duplicates()\n",
    "    df_genus.reset_index(drop=True, inplace=True)\n",
    "    df_genus[\"class\"] = [\"class_\" + str(i) for i in df_genus.index.tolist()]\n",
    "\n",
    "    ## if a class contains < 5 sequences, define as \"others\" class\n",
    "    df_virus = df_virus.merge(df_genus, on=[\"host_label\", \"Genus\"], how=\"left\")\n",
    "    for class_name in df_virus[\"class\"].unique():\n",
    "        df = df_virus[df_virus[\"class\"] == class_name]\n",
    "        if df.shape[0] < 5:\n",
    "            df_virus.loc[df_virus[\"class\"] == class_name, \"class\"] = \"others\"\n",
    "        else:\n",
    "            pass\n",
    "\n",
    "    ## if a class contains only either human or animal-infecting virus, define as \"others\" class\n",
    "    for genus in df_virus[\"Genus\"].unique():\n",
    "        df = df_virus[df_virus[\"Genus\"] == genus]\n",
    "        if len(df[\"host_label\"].unique()) == 1:\n",
    "            df_virus.loc[df_virus[\"Genus\"] == genus, \"class\"] = \"others\"\n",
    "    \n",
    "#     print (df_virus.groupby([\"class\", \"host_label\"]).count())\n",
    "    \n",
    "#     df_before = df_virus.copy()\n",
    "    \n",
    "    ## if a class contains < 5 sequences, exclude from the dataset\n",
    "    for class_name in df_virus[\"class\"].unique():\n",
    "        if df_virus[df_virus[\"class\"] == class_name].shape[0] < 5:\n",
    "            df_virus = df_virus[df_virus[\"class\"] != class_name]\n",
    "#             print (virus, c, df_before.shape[0], df_virus.shape[0])\n",
    "    \n",
    "    ## label information for five-fold stratified cross-validation\n",
    "    Y = df_virus[\"class\"].values.tolist()\n",
    "\n",
    "    #Y = df_virus[\"host_label\"].values.tolist()\n",
    "    #groups = df_virus[\"Genus\"].values.tolist()\n",
    "\n",
    "#     sgkf = StratifiedGroupKFold(n_splits=5, shuffle = True, random_state=42)\n",
    "    skf = StratifiedKFold(n_splits=5, shuffle = False, random_state=42)\n",
    "\n",
    "    fold_d = {}\n",
    "    fold_index_d = {}\n",
    "    n = 0\n",
    "    for train_index, test_index in skf.split(X, Y):\n",
    "        X_train = [x for i,x in enumerate(X) if i in train_index]\n",
    "        Y_train = [y for i,y in enumerate(Y) if i in train_index]\n",
    "\n",
    "        df_virus[\"dataset\"] = \"train\"\n",
    "        df_virus.loc[test_index, \"dataset\"] = \"test\"\n",
    "\n",
    "        X_train_1, X_eval_1, Y_train_1, Y_eval_1 = train_test_split(X_train, Y_train, test_size=0.2, stratify=Y_train, random_state=42)\n",
    "\n",
    "        df_virus.loc[df_virus[\"virus_ID\"].isin(X_eval_1), \"dataset\"] = \"eval\"\n",
    "\n",
    "        col_l = list(set(df_virus.columns.tolist()) & set(df_dataset.columns.tolist()))\n",
    "        df_acc = df_virus.merge(df_dataset, how=\"right\", on=col_l)\n",
    "        df_acc.reset_index(inplace=True)\n",
    "\n",
    "        df_train = df_acc[df_acc[\"dataset\"] == \"train\"]\n",
    "        df_eval = df_acc[df_acc[\"dataset\"] == \"eval\"]\n",
    "        df_test = df_acc[df_acc[\"dataset\"] == \"test\"]\n",
    "\n",
    "        train_acc = df_train[\"query\"].values.tolist()\n",
    "        eval_acc = df_eval[\"query\"].values.tolist()\n",
    "        test_acc = df_test[\"query\"].values.tolist()\n",
    "\n",
    "        train_index_l = df_train.index.tolist()\n",
    "        eval_index_l = df_eval.index.tolist()\n",
    "        test_index_l = df_test.index.tolist()\n",
    "\n",
    "        fold_d[n] = [train_acc, eval_acc, test_acc]\n",
    "        fold_index_d[n] = [train_index_l, eval_index_l, test_index_l]\n",
    "        #print (len(X_train), len(X_eval), len(X_test))\n",
    "\n",
    "#         n += 1\n",
    "    \n",
    "    return (fold_d, fold_index_d)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "start_time": "2023-03-22T03:31:27.719Z"
    }
   },
   "source": [
    "## visualization of fold data splitting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-09T02:15:37.307852Z",
     "start_time": "2023-11-09T02:15:37.302734Z"
    }
   },
   "outputs": [],
   "source": [
    "def confirm_fold_data(df_dataset, virus, fold_index_d):\n",
    "    df = df_dataset[df_dataset[\"Family\"] == virus]\n",
    "    df = df.sort_values([\"Genus\", \"host_label\"])\n",
    "    df.reset_index(inplace=True)\n",
    "    \n",
    "    n = 0\n",
    "    for fold, data in fold_index_d.items():\n",
    "        fold_name = \"fold\" + str(fold)\n",
    "        fold_l = [0] * len(df.index.tolist())\n",
    "        for i in df.index.tolist():\n",
    "            if i in data[0]:\n",
    "                fold_l[i] = \"train\"\n",
    "            elif i in data[1]:\n",
    "                fold_l[i] = \"eval\"\n",
    "            elif i in data[2]:\n",
    "                fold_l[i] = \"test\"\n",
    "            else:\n",
    "                print (i)\n",
    "\n",
    "        df[fold_name] = fold_l\n",
    "\n",
    "    df_fold = df[[\"Genus\", \"host_label\", \"fold0\", \"fold1\", \"fold2\", \"fold3\", \"fold4\"]]\n",
    "    df_count = df_fold.groupby([\"Genus\", \"host_label\"], dropna=False).count()\n",
    "    \n",
    "    return (df_fold, df_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-09T02:15:37.448806Z",
     "start_time": "2023-11-09T02:15:37.445279Z"
    }
   },
   "outputs": [],
   "source": [
    "# def visualize_dataset(output_dict, virus, df_fold):\n",
    "#     import catheat\n",
    "    \n",
    "#     #fig = plt.figure(figsize=(5,12))\n",
    "#     fig, axes = plt.subplots()\n",
    "    \n",
    "#     color_d = {\"train\":\"lightgray\", \"eval\":\"grey\", \"test\":\"black\", \"True\": \"blue\", \"False\": \"orange\"}\n",
    "\n",
    "#     catheat.heatmap(df_fold, cmap=color_d, palette=\"Set2\", leg_pos = \"right\", leg_kws={\"ncol\": 2})\n",
    "#     #plt.tight_layout()\n",
    "    \n",
    "#     plt.savefig(\"/\".join(map(str, [output_dict, virus, virus + \".png\"])))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## フォルダ作成"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-09T02:15:37.739584Z",
     "start_time": "2023-11-09T02:15:37.737327Z"
    }
   },
   "outputs": [],
   "source": [
    "# ##\n",
    "\n",
    "# def make_dir(path_l):\n",
    "#     dir_path = \"/\".join(path_l)\n",
    "    \n",
    "#     if os.path.exists(dir_path):\n",
    "#         pass\n",
    "#     else:\n",
    "#         os.mkdir(dir_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# データセットごとのリード配列の数"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-09T02:15:38.042106Z",
     "start_time": "2023-11-09T02:15:38.038881Z"
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# ##\n",
    "\n",
    "# def print_dataset_size(virus_family, key, fold, df_token):\n",
    "\n",
    "#     print (virus_family, key, \"human_virus_read\", \"animal_virus_read\")\n",
    "#     df_human = df_token[df_token[\"label\"] == 1]\n",
    "#     df_animal = df_token[df_token[\"label\"] == 0]\n",
    "    \n",
    "#     print (virus_family, key, fold, df_human.shape[0], df_animal.shape[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Main"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 配列データ読み込み"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-09T02:15:38.493482Z",
     "start_time": "2023-11-09T02:15:38.490122Z"
    }
   },
   "outputs": [],
   "source": [
    "##\n",
    "\n",
    "def read_seq_f(seq_f, out_dict):\n",
    "#     seq_f = out_dict + virus_family + \"/\" + virus_family + \".slide.250bp.fasta\"\n",
    "\n",
    "    seq_all_l = []\n",
    "    for n, line in enumerate(open(seq_f)):\n",
    "        line = line.strip()\n",
    "        if \">\" in line:\n",
    "            if n == 0:\n",
    "                pass\n",
    "            else:\n",
    "                seq_all_l.append(seq_l)\n",
    "\n",
    "            header = line.replace(\">\",\"\")\n",
    "            seq = \"\"\n",
    "            seq_l = [header, seq]\n",
    "\n",
    "        else:\n",
    "            seq = line\n",
    "            seq_l[1] += seq\n",
    "    \n",
    "    ## to DataFrame\n",
    "    df_seq = pd.DataFrame(seq_all_l, columns=[\"seqid\", \"seq\"])\n",
    "    df_seq[\"query\"] = [s.split(\"_sliding\")[0] for s in df_seq[\"seqid\"]]\n",
    "    \n",
    "    return (df_seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## tokenizer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-09T02:15:38.807701Z",
     "start_time": "2023-11-09T02:15:38.803138Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Pandarallel will run on 4 workers.\n",
      "INFO: Pandarallel will use standard multiprocessing data transfer (pipe) to transfer data between the main process and workers.\n"
     ]
    }
   ],
   "source": [
    "##\n",
    "from skbio import Sequence\n",
    "\n",
    "## 並列化\n",
    "from pandarallel import pandarallel\n",
    "pandarallel.initialize(nb_workers=4)\n",
    "\n",
    "def split_kmer(seq):\n",
    "    kmer_l = [k for k in Sequence(seq).iter_kmers(4, overlap=True)]\n",
    "    kmer = \" \".join(map(str, kmer_l))\n",
    "    return (kmer)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 宿主ラベルの追加"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-09T02:15:39.130005Z",
     "start_time": "2023-11-09T02:15:39.127319Z"
    }
   },
   "outputs": [],
   "source": [
    "def add_infectivity_label(df_dataset, df_seq):\n",
    "    df_label = df_dataset.loc[:, [\"query\", \"host_label\"]]\n",
    "    df_seq_1 = df_seq.merge(df_label, on = \"query\", how=\"left\")\n",
    "    df_seq_2 = df_seq_1.loc[:, [\"seqid\", \"host_label\", \"sequence\", \"query\"]]\n",
    "    return (df_seq_2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 既知・未知ウイルスの分割\n",
    "\n",
    "### 2018年より前後に分けて、ヒトウイルス or 動物ウイルスが20配列以上なら、未知ウイルスを取りおく\n",
    "\n",
    "### Orthomyxoviridae：未知ウイルスはマニュアルで集めないと厳しい可能性？"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-09T02:15:39.458486Z",
     "start_time": "2023-11-09T02:15:39.453281Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1721 293\n"
     ]
    }
   ],
   "source": [
    "df_known = df_dataset[df_dataset[\"collection_year\"] < 2018]\n",
    "df_unknown = df_dataset[df_dataset[\"collection_year\"] >= 2018]\n",
    "\n",
    "print (df_known.shape[0], df_unknown.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-09T02:15:39.613614Z",
     "start_time": "2023-11-09T02:15:39.609974Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(636, 41)"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_known[df_known[\"host_label\"] == 1].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-09T02:15:39.772901Z",
     "start_time": "2023-11-09T02:15:39.768926Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1085, 41)"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_known[df_known[\"host_label\"] == 0].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 【確認】Main"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-09T02:16:15.438860Z",
     "start_time": "2023-11-09T02:15:40.081719Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                    virus_ID  Genus\n",
      "class   host_label                 \n",
      "class_0 0                 13      0\n",
      "Reoviridae known_train human_virus_read animal_virus_read\n",
      "Reoviridae known_train 0 0 833\n",
      "Reoviridae known_eval human_virus_read animal_virus_read\n",
      "Reoviridae known_eval 0 0 146\n",
      "Reoviridae known_pred human_virus_read animal_virus_read\n",
      "Reoviridae known_pred 0 0 398\n",
      "Reoviridae known_train human_virus_read animal_virus_read\n",
      "Reoviridae known_train 1 0 864\n",
      "Reoviridae known_eval human_virus_read animal_virus_read\n",
      "Reoviridae known_eval 1 0 146\n",
      "Reoviridae known_pred human_virus_read animal_virus_read\n",
      "Reoviridae known_pred 1 0 367\n",
      "Reoviridae known_train human_virus_read animal_virus_read\n",
      "Reoviridae known_train 2 0 984\n",
      "Reoviridae known_eval human_virus_read animal_virus_read\n",
      "Reoviridae known_eval 2 0 79\n",
      "Reoviridae known_pred human_virus_read animal_virus_read\n",
      "Reoviridae known_pred 2 0 314\n",
      "Reoviridae known_train human_virus_read animal_virus_read\n",
      "Reoviridae known_train 3 0 984\n",
      "Reoviridae known_eval human_virus_read animal_virus_read\n",
      "Reoviridae known_eval 3 0 310\n",
      "Reoviridae known_pred human_virus_read animal_virus_read\n",
      "Reoviridae known_pred 3 0 83\n",
      "Reoviridae known_train human_virus_read animal_virus_read\n",
      "Reoviridae known_train 4 0 852\n",
      "Reoviridae known_eval human_virus_read animal_virus_read\n",
      "Reoviridae known_eval 4 0 310\n",
      "Reoviridae known_pred human_virus_read animal_virus_read\n",
      "Reoviridae known_pred 4 0 215\n"
     ]
    }
   ],
   "source": [
    "## Main\n",
    "\n",
    "kmer=4\n",
    "token=250\n",
    "out_dict = \"result/20230404/\"\n",
    "out_dict_1 = \"result/20230404\"\n",
    "\n",
    "for i,row in virus_dataset_d.iterrows():\n",
    "    virus_family = row[\"virus\"]\n",
    "    \n",
    "    if virus_family == \"Reoviridae\":\n",
    "    \n",
    "        # データセットの読み込み\n",
    "        dataset_f = row[\"virus_ID_file\"]\n",
    "        df_dataset = pd.read_csv(dataset_f, sep = \"\\t\", index_col = 0)\n",
    "\n",
    "        df_known = df_dataset[(df_dataset[\"collection_year\"] < 2018) & (df_dataset[\"Family\"] == virus_family)]\n",
    "\n",
    "        if \"query\" not in df_known.columns.tolist():\n",
    "            df_known[\"query\"] = [acc.split(\".\")[0] for acc in df_known[\"Accession\"]]\n",
    "\n",
    "        # 配列読み込み\n",
    "        df_seq = read_seq_f(virus_family, out_dict)\n",
    "        # seqを大文字に変換\n",
    "        df_seq[\"seq\"] = df_seq[\"seq\"].str.upper()\n",
    "        # kmerに分割\n",
    "        df_seq[\"sequence\"] = df_seq[\"seq\"].parallel_map(split_kmer)\n",
    "        # ラベル追加\n",
    "        df_token = add_infectivity_label(df_known, df_seq)\n",
    "\n",
    "        # 5-cross fold val datasetの作成\n",
    "        fold_d, fold_index_d = five_cross_fold_val(df_known, virus_family)\n",
    "\n",
    "        # ウイルス属ごとの配列数\n",
    "        df_fold, df_count = confirm_fold_data(df_known, virus_family, fold_index_d)\n",
    "\n",
    "        # データセット分割状況の可視化\n",
    "    #     visualize_dataset(out_dict_1, virus_family, df_fold)\n",
    "\n",
    "        # foldごとに出力\n",
    "        for fold, data in fold_d.items():\n",
    "            fold_name = \"fold_\" + str(fold)\n",
    "\n",
    "            train_acc_l = data[0]\n",
    "            eval_acc_l = data[1]\n",
    "            test_acc_l = data[2]\n",
    "\n",
    "            seq_d = {\"known_train\": [[out_dict_1, virus_family, \"ft_known_4_250_\" + fold_name, \"train.tsv\"], train_acc_l],\n",
    "                 \"known_eval\": [[out_dict_1, virus_family, \"ft_known_4_250_\" + fold_name, \"dev.tsv\"], eval_acc_l],\n",
    "                 \"known_pred\": [[out_dict_1, virus_family, \"pred_known_4_250_\" + fold_name, \"dev.tsv\"], test_acc_l]}\n",
    "\n",
    "            for key, data_1 in seq_d.items():\n",
    "                dict_l = data_1[0]\n",
    "                acc_l = data_1[1]\n",
    "                df = df_token[df_token[\"query\"].isin(acc_l)]\n",
    "\n",
    "                df_output = df.rename(columns={\"host_label\":\"label\"})\n",
    "                df_output = df_output.loc[:, [\"sequence\", \"label\", \"seqid\"]]\n",
    "                df_output = df_output.astype({\"sequence\": str, \"label\": int, \"seqid\":str})\n",
    "\n",
    "                make_dir(dict_l[:-1])\n",
    "                output_f = \"/\".join(map(str, dict_l))\n",
    "\n",
    "    #             df_output.to_csv(output_f, sep=\"\\t\", index=False)\n",
    "                #print (df_output.shape)\n",
    "\n",
    "                # データセットの内訳\n",
    "#                 print_dataset_size(virus_family, key, fold, df_output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-09T02:16:15.448733Z",
     "start_time": "2023-11-09T02:16:15.441781Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>fold0</th>\n",
       "      <th>fold1</th>\n",
       "      <th>fold2</th>\n",
       "      <th>fold3</th>\n",
       "      <th>fold4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genus</th>\n",
       "      <th>host_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NaN</th>\n",
       "      <th>0</th>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  fold0  fold1  fold2  fold3  fold4\n",
       "Genus host_label                                   \n",
       "NaN   0              75     75     75     75     75"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-08T06:39:25.401213Z",
     "start_time": "2023-08-08T06:39:25.397122Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/usr/bin/env python\n",
      "import sys, re\n",
      "argvs = sys.argv\n",
      "sys.setrecursionlimit(10000000)\n",
      "\n",
      "import os\n",
      "import collections\n",
      "import numpy as np\n",
      "import pandas as pd\n",
      "import array\n",
      "pd.set_option(\"display.max_colwidth\", 200)\n",
      "pd.set_option(\"display.max_columns\", 200)\n",
      "pd.set_option(\"display.max_rows\", 300)\n",
      "\n",
      "import glob\n",
      "\n",
      "## visualize\n",
      "import matplotlib\n",
      "import matplotlib.pyplot as plt\n",
      "import seaborn as sns\n",
      "sns.set()\n",
      "\n",
      "plt.rcParams['font.family'] = \"Arial\"\n",
      "virus_l = [\"Orthomyxoviridae\",  \"Hantaviridae\", \"Nairoviridae\", \"Peribunyaviridae\", \"Pneumoviridae\",\n",
      "           \"Phenuiviridae\", \"Arenaviridae\", \"Reoviridae\"]\n",
      "%%bash\n",
      "\n",
      "virus_family=\"Arenaviridae\"\n",
      "\n",
      "output_dict=\"result/20230718\"\n",
      "acc_file=\"data/20230404_genbank/\"${virus_family}\".curated.acc.txt\"\n",
      "seq_f=\"data/20230404_genbank/\"${virus_family}\".curated\"\n",
      "\n",
      "mkdir ${output_dict}/${virus_family}\n",
      "\n",
      "python \\\n",
      "    /Users/junna/Desktop/note/script/entrez_fasta_20230127.py \\\n",
      "    ${acc_file} \\\n",
      "    ${seq_f}.gb\n",
      "    \n",
      "seqret -snucleotide1 \\\n",
      "    -sequence ${seq_f}.gb \\\n",
      "    -outseq ${seq_f}.fasta\n",
      "    \n",
      "${acc_file} | wc -l\n",
      "seqkit stats ${seq_f}.fasta\n",
      "virus_l = [\"Orthomyxoviridae\",  \"Hantaviridae\", \"Nairoviridae\", \"Peribunyaviridae\", \"Pneumoviridae\",\n",
      "           \"Phenuiviridae\", \"Arenaviridae\", \"Reoviridae\"]\n",
      "%%bash\n",
      "\n",
      "virus_family=\"Arenaviridae\"\n",
      "\n",
      "output_dict=\"result/20230808\"\n",
      "acc_file=\"data/20230404_genbank/\"${virus_family}\".curated.acc.txt\"\n",
      "seq_f=\"data/20230404_genbank/\"${virus_family}\".curated\"\n",
      "\n",
      "cat ${seq_f}.fasta | \\\n",
      "    seqkit replace -s -p \"[^atcguATCGU]\" -r \"-\" \\\n",
      "    > ${seq_f}.QC.fasta\n",
      "\n",
      "cat ${seq_f}.QC.fasta | \\\n",
      "    seqkit sliding -s 125 -W 250 --greedy | \\\n",
      "    seqkit seq -m 100 \\\n",
      "    > ${output_dict}/${virus_family}/${virus_family}.slide.250bp.fasta\n",
      "\n",
      "cd-hit-est \\\n",
      "    -n 10 -c 1.0 -d 0 -M 10000 -aS 1.0 \\\n",
      "    -i ${output_dict}/${virus_family}/${virus_family}.slide.250bp.fasta \\\n",
      "    -o ${output_dict}/${virus_family}/${virus_family}.slide.250bp.cdhit100.fasta\n",
      "\n",
      "seqkit stats ${output_dict}/${virus_family}/${virus_family}.slide.250bp.cdhit100.fasta\n",
      "seqkit stats ${output_dict}/${virus_family}/${virus_family}.slide.250bp.fasta\n",
      "%%bash\n",
      "\n",
      "virus_family=\"Arenaviridae\"\n",
      "\n",
      "output_dict=\"result/20230808\"\n",
      "acc_file=\"data/20230404_genbank/\"${virus_family}\".curated.acc.txt\"\n",
      "seq_f=\"data/20230404_genbank/\"${virus_family}\".curated\"\n",
      "\n",
      "cat ${seq_f}.fasta | \\\n",
      "    seqkit replace -s -p \"[^atcguATCGU]\" -r \"-\" \\\n",
      "    > ${seq_f}.QC.fasta\n",
      "\n",
      "cat ${seq_f}.QC.fasta | \\\n",
      "    seqkit sliding -s 125 -W 250 --greedy | \\\n",
      "    seqkit seq -m 100 \\\n",
      "    > ${output_dict}/${virus_family}/${virus_family}.slide.250bp.fasta\n",
      "\n",
      "cd-hit-est \\\n",
      "    -n 10 -c 1.0 -d 0 -M 10000 -aS 1.0 \\\n",
      "    -i ${output_dict}/${virus_family}/${virus_family}.slide.250bp.fasta \\\n",
      "    -o ${output_dict}/${virus_family}/${virus_family}.slide.250bp.cdhit100.fasta\n",
      "\n",
      "seqkit stats ${output_dict}/${virus_family}/${virus_family}.slide.250bp.cdhit100.fasta\n",
      "seqkit stats ${output_dict}/${virus_family}/${virus_family}.slide.250bp.fasta\n",
      "\n",
      "seqkit stats ${seq_f}.QC.fasta\n",
      "virus_l = [\"Arenaviridae\"]\n",
      "virus = virus_l[0]\n",
      "\n",
      "dataset_f = \"data/20230404_genbank/\" + virus + \".curated.csv\"\n",
      "df_dataset = pd.read_csv(dataset_f, sep = \"\\t\", index_col = 0)\n",
      "df_dataset.head(3)\n",
      "df_dataset[\"Genus\"].unique()\n",
      "df_dataset.shape\n",
      "df_dataset = df_dataset[~(df_dataset[\"Host\"].isnull())]\n",
      "df_dataset.shape\n",
      "print (df_dataset[df_dataset[\"host_label\"] == 1].shape, df_dataset[df_dataset[\"host_label\"] == 0].shape)\n",
      "def five_cross_fold_val(df_dataset, virus):\n",
      "    from sklearn.model_selection import StratifiedGroupKFold, StratifiedKFold\n",
      "    from operator import itemgetter\n",
      "    from sklearn.model_selection import train_test_split\n",
      "\n",
      "    df_all = df_dataset[df_dataset[\"Family\"] == virus]\n",
      "    df_all = df_all.sort_values([\"Genus\", \"host_label\"])\n",
      "    df_all.reset_index(drop=True, inplace=True)\n",
      "    \n",
      "    # virus_IDごとにデータセットに分ける\n",
      "    if \"virus_ID\" not in df_all.columns.tolist():\n",
      "        df_all[\"virus_ID\"] = df_all[\"Accession\"].copy()\n",
      "    \n",
      "    # virus_IDごとにデータセットに分ける\n",
      "    df_virus = df_all.loc[:, [\"virus_ID\", \"host_label\", \"Genus\"]].drop_duplicates()\n",
      "    df_virus.reset_index(drop=True, inplace=True)\n",
      "    X = df_virus[\"virus_ID\"].values.tolist()\n",
      "\n",
      "    ## ウイルス属と宿主の組み合わせでクラスを定義づける\n",
      "    df_genus = df_all.loc[:, [\"host_label\", \"Genus\"]].drop_duplicates()\n",
      "    df_genus.reset_index(drop=True, inplace=True)\n",
      "    df_genus[\"class\"] = [\"class_\" + str(i) for i in df_genus.index.tolist()]\n",
      "\n",
      "    ## クラスに対する配列数が少なすぎる場合は「others」としてまとめる\n",
      "    df_virus = df_virus.merge(df_genus, on=[\"host_label\", \"Genus\"], how=\"left\")\n",
      "    for class_name in df_virus[\"class\"].unique():\n",
      "        df = df_virus[df_virus[\"class\"] == class_name]\n",
      "        if df.shape[0] < 5:\n",
      "            df_virus.loc[df_virus[\"class\"] == class_name, \"class\"] = \"others\"\n",
      "        else:\n",
      "            pass\n",
      "\n",
      "    ## 1つのウイルス属にヒト・動物のどちらかのウイルスしか存在しない場合は、クラスを「others」とする\n",
      "    for genus in df_virus[\"Genus\"].unique():\n",
      "        df = df_virus[df_virus[\"Genus\"] == genus]\n",
      "        if len(df[\"host_label\"].unique()) == 1:\n",
      "            df_virus.loc[df_virus[\"Genus\"] == genus, \"class\"] = \"others\"\n",
      "    \n",
      "    print (df_virus.groupby(\"class\"))\n",
      "    \n",
      "    ## ラベル\n",
      "    Y = df_virus[\"class\"].values.tolist()\n",
      "\n",
      "    #Y = df_virus[\"host_label\"].values.tolist()\n",
      "    #groups = df_virus[\"Genus\"].values.tolist()\n",
      "\n",
      "    sgkf = StratifiedGroupKFold(n_splits=5, shuffle = True)\n",
      "    skf = StratifiedKFold(n_splits=5, shuffle = False)\n",
      "\n",
      "    fold_d = {}\n",
      "    fold_index_d = {}\n",
      "    n = 0\n",
      "    for train_index, test_index in skf.split(X, Y):\n",
      "        X_train = [x for i,x in enumerate(X) if i in train_index]\n",
      "        Y_train = [y for i,y in enumerate(Y) if i in train_index]\n",
      "\n",
      "        df_virus[\"dataset\"] = 0\n",
      "        df_virus.loc[train_index, \"dataset\"] = \"train\"\n",
      "        df_virus.loc[test_index, \"dataset\"] = \"test\"\n",
      "\n",
      "        X_train_1, X_eval_1, Y_train_1, Y_eval_1 = train_test_split(X_train, Y_train, test_size=0.2, stratify=Y_train)\n",
      "\n",
      "        df_virus.loc[df_virus[\"virus_ID\"].isin(X_eval_1), \"dataset\"] = \"eval\"\n",
      "\n",
      "        col_l = list(set(df_virus.columns.tolist()) & set(df_all.columns.tolist()))\n",
      "        df_acc = df_virus.merge(df_all, how=\"right\", on=col_l)\n",
      "        df_acc.reset_index(inplace=True)\n",
      "\n",
      "        df_train = df_acc[df_acc[\"dataset\"] == \"train\"]\n",
      "        df_eval = df_acc[df_acc[\"dataset\"] == \"eval\"]\n",
      "        df_test = df_acc[df_acc[\"dataset\"] == \"test\"]\n",
      "\n",
      "        train_acc = df_train[\"query\"].values.tolist()\n",
      "        eval_acc = df_eval[\"query\"].values.tolist()\n",
      "        test_acc = df_test[\"query\"].values.tolist()\n",
      "\n",
      "        train_index_l = df_train.index.tolist()\n",
      "        eval_index_l = df_eval.index.tolist()\n",
      "        test_index_l = df_test.index.tolist()\n",
      "\n",
      "        fold_d[n] = [train_acc, eval_acc, test_acc]\n",
      "        fold_index_d[n] = [train_index_l, eval_index_l, test_index_l]\n",
      "        #print (len(X_train), len(X_eval), len(X_test))\n",
      "\n",
      "        n += 1\n",
      "    \n",
      "    return (fold_d, fold_index_d)\n",
      "def confirm_fold_data(df_dataset, virus, fold_index_d):\n",
      "    df = df_dataset[df_dataset[\"Family\"] == virus]\n",
      "    df = df.sort_values([\"Genus\", \"host_label\"])\n",
      "    df.reset_index(inplace=True)\n",
      "    \n",
      "    n = 0\n",
      "    for fold, data in fold_index_d.items():\n",
      "        fold_name = \"fold\" + str(fold)\n",
      "        fold_l = [0] * len(df.index.tolist())\n",
      "        for i in df.index.tolist():\n",
      "            if i in data[0]:\n",
      "                fold_l[i] = \"train\"\n",
      "            elif i in data[1]:\n",
      "                fold_l[i] = \"eval\"\n",
      "            elif i in data[2]:\n",
      "                fold_l[i] = \"test\"\n",
      "            else:\n",
      "                print (i)\n",
      "\n",
      "        df[fold_name] = fold_l\n",
      "\n",
      "    df_fold = df[[\"Genus\", \"host_label\", \"fold0\", \"fold1\", \"fold2\", \"fold3\", \"fold4\"]]\n",
      "    df_count = df_fold.groupby([\"Genus\", \"host_label\"], dropna=False).count()\n",
      "    \n",
      "    return (df_fold, df_count)\n",
      "def visualize_dataset(output_dict, virus, df_fold):\n",
      "    import catheat\n",
      "    \n",
      "    #fig = plt.figure(figsize=(5,12))\n",
      "    fig, axes = plt.subplots()\n",
      "    \n",
      "    color_d = {\"train\":\"lightgray\", \"eval\":\"grey\", \"test\":\"black\", \"True\": \"blue\", \"False\": \"orange\"}\n",
      "\n",
      "    catheat.heatmap(df_fold, cmap=color_d, palette=\"Set2\", leg_pos = \"right\", leg_kws={\"ncol\": 2})\n",
      "    #plt.tight_layout()\n",
      "    \n",
      "    plt.savefig(\"/\".join(map(str, [output_dict, virus, virus + \".png\"])))\n",
      "##\n",
      "\n",
      "def make_dir(path_l):\n",
      "    dir_path = \"/\".join(path_l)\n",
      "    \n",
      "    if os.path.exists(dir_path):\n",
      "        pass\n",
      "    else:\n",
      "        os.mkdir(dir_path)\n",
      "##\n",
      "\n",
      "def print_dataset_size(virus_family, key, fold, df_token):\n",
      "\n",
      "    print (virus_family, key, \"human_virus_read\", \"animal_virus_read\")\n",
      "    df_human = df_token[df_token[\"label\"] == 1]\n",
      "    df_animal = df_token[df_token[\"label\"] == 0]\n",
      "    \n",
      "    print (virus_family, key, fold, df_human.shape[0], df_animal.shape[0])\n",
      "##\n",
      "\n",
      "def read_seq_f(virus_family, out_dict):\n",
      "    seq_f = out_dict + virus_family + \"/\" + virus_family + \".slide.250bp.fasta\"\n",
      "\n",
      "    seq_all_l = []\n",
      "    for n, line in enumerate(open(seq_f)):\n",
      "        line = line.strip()\n",
      "        if \">\" in line:\n",
      "            if n == 0:\n",
      "                pass\n",
      "            else:\n",
      "                seq_all_l.append(seq_l)\n",
      "\n",
      "            header = line.replace(\">\",\"\")\n",
      "            seq = \"\"\n",
      "            seq_l = [header, seq]\n",
      "\n",
      "        else:\n",
      "            seq = line\n",
      "            seq_l[1] += seq\n",
      "    \n",
      "    ## to DataFrame\n",
      "    df_seq = pd.DataFrame(seq_all_l, columns=[\"seqid\", \"seq\"])\n",
      "    df_seq[\"query\"] = [s.split(\"_sliding\")[0] for s in df_seq[\"seqid\"]]\n",
      "    \n",
      "    return (df_seq)\n",
      "##\n",
      "from skbio import Sequence\n",
      "\n",
      "## 並列化\n",
      "from pandarallel import pandarallel\n",
      "pandarallel.initialize()\n",
      "\n",
      "def split_kmer(seq):\n",
      "    kmer_l = [k for k in Sequence(seq).iter_kmers(4, overlap=True)]\n",
      "    kmer = \" \".join(map(str, kmer_l))\n",
      "    return (kmer)\n",
      "def add_infectivity_label(df_dataset, df_seq):\n",
      "    df_label = df_dataset.loc[:, [\"query\", \"host_label\"]]\n",
      "    df_seq_1 = df_seq.merge(df_label, on = \"query\", how=\"left\")\n",
      "    df_seq_2 = df_seq_1.loc[:, [\"seqid\", \"host_label\", \"sequence\", \"query\"]]\n",
      "    return (df_seq_2)\n",
      "df_known = df_dataset[df_dataset[\"collection_year\"] < 2018]\n",
      "df_unknown = df_dataset[df_dataset[\"collection_year\"] >= 2018]\n",
      "\n",
      "print (df_known.shape[0], df_unknown.shape[0])\n",
      "df_known[df_known[\"host_label\"] == 1].shape\n",
      "## Main\n",
      "\n",
      "kmer=4\n",
      "token=250\n",
      "out_dict = \"result/20230718/\"\n",
      "out_dict_1 = \"result/20230718\"\n",
      "\n",
      "for virus_family in virus_l:\n",
      "    # 配列読み込み\n",
      "    df_seq = read_seq_f(virus_family, out_dict)\n",
      "    # seqを大文字に変換\n",
      "    df_seq[\"seq\"] = df_seq[\"seq\"].str.upper()\n",
      "    # kmerに分割\n",
      "    df_seq[\"sequence\"] = df_seq[\"seq\"].parallel_map(split_kmer)\n",
      "    # ラベル追加\n",
      "    df_token = add_infectivity_label(df_known, df_seq)\n",
      "    \n",
      "    # 5-cross fold val datasetの作成\n",
      "    fold_d, fold_index_d = five_cross_fold_val(df_known, virus_family)\n",
      "    \n",
      "    # ウイルス属ごとの配列数\n",
      "    df_fold, df_count = confirm_fold_data(df_known, virus_family, fold_index_d)\n",
      "    \n",
      "    # データセット分割状況の可視化\n",
      "    visualize_dataset(out_dict_1, virus_family, df_fold)\n",
      "    \n",
      "    # foldごとに出力\n",
      "    for fold, data in fold_d.items():\n",
      "        fold_name = \"fold_\" + str(fold)\n",
      "        \n",
      "        train_acc_l = data[0]\n",
      "        eval_acc_l = data[1]\n",
      "        test_acc_l = data[2]\n",
      "    \n",
      "        seq_d = {\"known_train\": [[out_dict_1, virus_family, \"ft_known_4_250_\" + fold_name, \"train.tsv\"], train_acc_l],\n",
      "             \"known_eval\": [[out_dict_1, virus_family, \"ft_known_4_250_\" + fold_name, \"dev.tsv\"], eval_acc_l],\n",
      "             \"known_pred\": [[out_dict_1, virus_family, \"pred_known_4_250_\" + fold_name, \"dev.tsv\"], test_acc_l]}\n",
      "\n",
      "        for key, data_1 in seq_d.items():\n",
      "            dict_l = data_1[0]\n",
      "            acc_l = data_1[1]\n",
      "            df = df_token[df_token[\"query\"].isin(acc_l)]\n",
      "            \n",
      "            df_output = df.rename(columns={\"host_label\":\"label\"})\n",
      "            df_output = df_output.loc[:, [\"sequence\", \"label\", \"seqid\"]]\n",
      "            df_output = df_output.astype({\"sequence\": str, \"label\": int, \"seqid\":str})\n",
      "\n",
      "            make_dir(dict_l[:-1])\n",
      "            output_f = \"/\".join(map(str, dict_l))\n",
      "\n",
      "            df_output.to_csv(output_f, sep=\"\\t\", index=False)\n",
      "            #print (df_output.shape)\n",
      "\n",
      "            # データセットの内訳\n",
      "            print_dataset_size(virus_family, key, fold, df_output)\n",
      "df_count\n",
      "#!/usr/bin/env python\n",
      "import sys, re\n",
      "argvs = sys.argv\n",
      "sys.setrecursionlimit(10000000)\n",
      "\n",
      "import os\n",
      "import collections\n",
      "import numpy as np\n",
      "import pandas as pd\n",
      "import array\n",
      "pd.set_option(\"display.max_colwidth\", 200)\n",
      "pd.set_option(\"display.max_columns\", 200)\n",
      "pd.set_option(\"display.max_rows\", 300)\n",
      "\n",
      "import glob\n",
      "\n",
      "## visualize\n",
      "import matplotlib\n",
      "import matplotlib.pyplot as plt\n",
      "import seaborn as sns\n",
      "sns.set()\n",
      "\n",
      "plt.rcParams['font.family'] = \"Arial\"\n",
      "virus_l = [\"Orthomyxoviridae\",  \"Hantaviridae\", \"Nairoviridae\", \"Peribunyaviridae\", \"Pneumoviridae\",\n",
      "           \"Phenuiviridae\", \"Arenaviridae\", \"Reoviridae\"]\n",
      "%%bash\n",
      "\n",
      "virus_family=\"Arenaviridae\"\n",
      "\n",
      "output_dict=\"result/20230808\"\n",
      "acc_file=\"data/20230404_genbank/\"${virus_family}\".curated.acc.txt\"\n",
      "seq_f=\"data/20230404_genbank/\"${virus_family}\".curated\"\n",
      "\n",
      "mkdir ${output_dict}/${virus_family}\n",
      "\n",
      "python \\\n",
      "    /Users/junna/Desktop/note/script/entrez_fasta_20230127.py \\\n",
      "    ${acc_file} \\\n",
      "    ${seq_f}.gb\n",
      "    \n",
      "seqret -snucleotide1 \\\n",
      "    -sequence ${seq_f}.gb \\\n",
      "    -outseq ${seq_f}.fasta\n",
      "    \n",
      "${acc_file} | wc -l\n",
      "seqkit stats ${seq_f}.fasta\n",
      "%%bash\n",
      "\n",
      "virus_family=\"Arenaviridae\"\n",
      "\n",
      "output_dict=\"result/20230808\"\n",
      "acc_file=\"data/20230404_genbank/\"${virus_family}\".curated.acc.txt\"\n",
      "seq_f=\"data/20230404_genbank/\"${virus_family}\".curated\"\n",
      "\n",
      "cat ${seq_f}.fasta | \\\n",
      "    seqkit replace -s -p \"[^atcguATCGU]\" -r \"-\" \\\n",
      "    > ${seq_f}.QC.fasta\n",
      "\n",
      "cat ${seq_f}.QC.fasta | \\\n",
      "    seqkit sliding -s 125 -W 250 --greedy | \\\n",
      "    seqkit seq -m 100 \\\n",
      "    > ${output_dict}/${virus_family}/${virus_family}.slide.250bp.fasta\n",
      "\n",
      "cd-hit-est \\\n",
      "    -n 10 -c 1.0 -d 0 -M 10000 -aS 1.0 \\\n",
      "    -i ${output_dict}/${virus_family}/${virus_family}.slide.250bp.fasta \\\n",
      "    -o ${output_dict}/${virus_family}/${virus_family}.slide.250bp.cdhit100.fasta\n",
      "\n",
      "seqkit stats ${output_dict}/${virus_family}/${virus_family}.slide.250bp.cdhit100.fasta\n",
      "seqkit stats ${output_dict}/${virus_family}/${virus_family}.slide.250bp.fasta\n",
      "\n",
      "seqkit stats ${seq_f}.QC.fasta\n",
      "virus_l = [\"Arenaviridae\"]\n",
      "virus = virus_l[0]\n",
      "\n",
      "dataset_f = \"data/20230404_genbank/\" + virus + \".curated.csv\"\n",
      "df_dataset = pd.read_csv(dataset_f, sep = \"\\t\", index_col = 0)\n",
      "df_dataset.head(3)\n",
      "df_dataset[\"Genus\"].unique()\n",
      "df_dataset.shape\n",
      "df_dataset = df_dataset[~(df_dataset[\"Host\"].isnull())]\n",
      "df_dataset.shape\n",
      "print (df_dataset[df_dataset[\"host_label\"] == 1].shape, df_dataset[df_dataset[\"host_label\"] == 0].shape)\n",
      "def five_cross_fold_val(df_dataset, virus):\n",
      "    from sklearn.model_selection import StratifiedGroupKFold, StratifiedKFold\n",
      "    from operator import itemgetter\n",
      "    from sklearn.model_selection import train_test_split\n",
      "\n",
      "    df_all = df_dataset[df_dataset[\"Family\"] == virus]\n",
      "    df_all = df_all.sort_values([\"Genus\", \"host_label\"])\n",
      "    df_all.reset_index(drop=True, inplace=True)\n",
      "    \n",
      "    # virus_IDごとにデータセットに分ける\n",
      "    if \"virus_ID\" not in df_all.columns.tolist():\n",
      "        df_all[\"virus_ID\"] = df_all[\"Accession\"].copy()\n",
      "    \n",
      "    # virus_IDごとにデータセットに分ける\n",
      "    df_virus = df_all.loc[:, [\"virus_ID\", \"host_label\", \"Genus\"]].drop_duplicates()\n",
      "    df_virus.reset_index(drop=True, inplace=True)\n",
      "    X = df_virus[\"virus_ID\"].values.tolist()\n",
      "\n",
      "    ## ウイルス属と宿主の組み合わせでクラスを定義づける\n",
      "    df_genus = df_all.loc[:, [\"host_label\", \"Genus\"]].drop_duplicates()\n",
      "    df_genus.reset_index(drop=True, inplace=True)\n",
      "    df_genus[\"class\"] = [\"class_\" + str(i) for i in df_genus.index.tolist()]\n",
      "\n",
      "    ## クラスに対する配列数が少なすぎる場合は「others」としてまとめる\n",
      "    df_virus = df_virus.merge(df_genus, on=[\"host_label\", \"Genus\"], how=\"left\")\n",
      "    for class_name in df_virus[\"class\"].unique():\n",
      "        df = df_virus[df_virus[\"class\"] == class_name]\n",
      "        if df.shape[0] < 5:\n",
      "            df_virus.loc[df_virus[\"class\"] == class_name, \"class\"] = \"others\"\n",
      "        else:\n",
      "            pass\n",
      "\n",
      "    ## 1つのウイルス属にヒト・動物のどちらかのウイルスしか存在しない場合は、クラスを「others」とする\n",
      "    for genus in df_virus[\"Genus\"].unique():\n",
      "        df = df_virus[df_virus[\"Genus\"] == genus]\n",
      "        if len(df[\"host_label\"].unique()) == 1:\n",
      "            df_virus.loc[df_virus[\"Genus\"] == genus, \"class\"] = \"others\"\n",
      "    \n",
      "    print (df_virus.groupby(\"class\"))\n",
      "    \n",
      "    ## ラベル\n",
      "    Y = df_virus[\"class\"].values.tolist()\n",
      "\n",
      "    #Y = df_virus[\"host_label\"].values.tolist()\n",
      "    #groups = df_virus[\"Genus\"].values.tolist()\n",
      "\n",
      "    sgkf = StratifiedGroupKFold(n_splits=5, shuffle = True)\n",
      "    skf = StratifiedKFold(n_splits=5, shuffle = False)\n",
      "\n",
      "    fold_d = {}\n",
      "    fold_index_d = {}\n",
      "    n = 0\n",
      "    for train_index, test_index in skf.split(X, Y):\n",
      "        X_train = [x for i,x in enumerate(X) if i in train_index]\n",
      "        Y_train = [y for i,y in enumerate(Y) if i in train_index]\n",
      "\n",
      "        df_virus[\"dataset\"] = 0\n",
      "        df_virus.loc[train_index, \"dataset\"] = \"train\"\n",
      "        df_virus.loc[test_index, \"dataset\"] = \"test\"\n",
      "\n",
      "        X_train_1, X_eval_1, Y_train_1, Y_eval_1 = train_test_split(X_train, Y_train, test_size=0.2, stratify=Y_train)\n",
      "\n",
      "        df_virus.loc[df_virus[\"virus_ID\"].isin(X_eval_1), \"dataset\"] = \"eval\"\n",
      "\n",
      "        col_l = list(set(df_virus.columns.tolist()) & set(df_all.columns.tolist()))\n",
      "        df_acc = df_virus.merge(df_all, how=\"right\", on=col_l)\n",
      "        df_acc.reset_index(inplace=True)\n",
      "\n",
      "        df_train = df_acc[df_acc[\"dataset\"] == \"train\"]\n",
      "        df_eval = df_acc[df_acc[\"dataset\"] == \"eval\"]\n",
      "        df_test = df_acc[df_acc[\"dataset\"] == \"test\"]\n",
      "\n",
      "        train_acc = df_train[\"query\"].values.tolist()\n",
      "        eval_acc = df_eval[\"query\"].values.tolist()\n",
      "        test_acc = df_test[\"query\"].values.tolist()\n",
      "\n",
      "        train_index_l = df_train.index.tolist()\n",
      "        eval_index_l = df_eval.index.tolist()\n",
      "        test_index_l = df_test.index.tolist()\n",
      "\n",
      "        fold_d[n] = [train_acc, eval_acc, test_acc]\n",
      "        fold_index_d[n] = [train_index_l, eval_index_l, test_index_l]\n",
      "        #print (len(X_train), len(X_eval), len(X_test))\n",
      "\n",
      "        n += 1\n",
      "    \n",
      "    return (fold_d, fold_index_d)\n",
      "def confirm_fold_data(df_dataset, virus, fold_index_d):\n",
      "    df = df_dataset[df_dataset[\"Family\"] == virus]\n",
      "    df = df.sort_values([\"Genus\", \"host_label\"])\n",
      "    df.reset_index(inplace=True)\n",
      "    \n",
      "    n = 0\n",
      "    for fold, data in fold_index_d.items():\n",
      "        fold_name = \"fold\" + str(fold)\n",
      "        fold_l = [0] * len(df.index.tolist())\n",
      "        for i in df.index.tolist():\n",
      "            if i in data[0]:\n",
      "                fold_l[i] = \"train\"\n",
      "            elif i in data[1]:\n",
      "                fold_l[i] = \"eval\"\n",
      "            elif i in data[2]:\n",
      "                fold_l[i] = \"test\"\n",
      "            else:\n",
      "                print (i)\n",
      "\n",
      "        df[fold_name] = fold_l\n",
      "\n",
      "    df_fold = df[[\"Genus\", \"host_label\", \"fold0\", \"fold1\", \"fold2\", \"fold3\", \"fold4\"]]\n",
      "    df_count = df_fold.groupby([\"Genus\", \"host_label\"], dropna=False).count()\n",
      "    \n",
      "    return (df_fold, df_count)\n",
      "def visualize_dataset(output_dict, virus, df_fold):\n",
      "    import catheat\n",
      "    \n",
      "    #fig = plt.figure(figsize=(5,12))\n",
      "    fig, axes = plt.subplots()\n",
      "    \n",
      "    color_d = {\"train\":\"lightgray\", \"eval\":\"grey\", \"test\":\"black\", \"True\": \"blue\", \"False\": \"orange\"}\n",
      "\n",
      "    catheat.heatmap(df_fold, cmap=color_d, palette=\"Set2\", leg_pos = \"right\", leg_kws={\"ncol\": 2})\n",
      "    #plt.tight_layout()\n",
      "    \n",
      "    plt.savefig(\"/\".join(map(str, [output_dict, virus, virus + \".png\"])))\n",
      "##\n",
      "\n",
      "def make_dir(path_l):\n",
      "    dir_path = \"/\".join(path_l)\n",
      "    \n",
      "    if os.path.exists(dir_path):\n",
      "        pass\n",
      "    else:\n",
      "        os.mkdir(dir_path)\n",
      "##\n",
      "\n",
      "def print_dataset_size(virus_family, key, fold, df_token):\n",
      "\n",
      "    print (virus_family, key, \"human_virus_read\", \"animal_virus_read\")\n",
      "    df_human = df_token[df_token[\"label\"] == 1]\n",
      "    df_animal = df_token[df_token[\"label\"] == 0]\n",
      "    \n",
      "    print (virus_family, key, fold, df_human.shape[0], df_animal.shape[0])\n",
      "##\n",
      "\n",
      "def read_seq_f(virus_family, out_dict):\n",
      "    seq_f = out_dict + virus_family + \"/\" + virus_family + \".slide.250bp.fasta\"\n",
      "\n",
      "    seq_all_l = []\n",
      "    for n, line in enumerate(open(seq_f)):\n",
      "        line = line.strip()\n",
      "        if \">\" in line:\n",
      "            if n == 0:\n",
      "                pass\n",
      "            else:\n",
      "                seq_all_l.append(seq_l)\n",
      "\n",
      "            header = line.replace(\">\",\"\")\n",
      "            seq = \"\"\n",
      "            seq_l = [header, seq]\n",
      "\n",
      "        else:\n",
      "            seq = line\n",
      "            seq_l[1] += seq\n",
      "    \n",
      "    ## to DataFrame\n",
      "    df_seq = pd.DataFrame(seq_all_l, columns=[\"seqid\", \"seq\"])\n",
      "    df_seq[\"query\"] = [s.split(\"_sliding\")[0] for s in df_seq[\"seqid\"]]\n",
      "    \n",
      "    return (df_seq)\n",
      "##\n",
      "from skbio import Sequence\n",
      "\n",
      "## 並列化\n",
      "from pandarallel import pandarallel\n",
      "pandarallel.initialize()\n",
      "\n",
      "def split_kmer(seq):\n",
      "    kmer_l = [k for k in Sequence(seq).iter_kmers(4, overlap=True)]\n",
      "    kmer = \" \".join(map(str, kmer_l))\n",
      "    return (kmer)\n",
      "def add_infectivity_label(df_dataset, df_seq):\n",
      "    df_label = df_dataset.loc[:, [\"query\", \"host_label\"]]\n",
      "    df_seq_1 = df_seq.merge(df_label, on = \"query\", how=\"left\")\n",
      "    df_seq_2 = df_seq_1.loc[:, [\"seqid\", \"host_label\", \"sequence\", \"query\"]]\n",
      "    return (df_seq_2)\n",
      "df_known = df_dataset[df_dataset[\"collection_year\"] < 2018]\n",
      "df_unknown = df_dataset[df_dataset[\"collection_year\"] >= 2018]\n",
      "\n",
      "print (df_known.shape[0], df_unknown.shape[0])\n",
      "df_known[df_known[\"host_label\"] == 1].shape\n",
      "## Main\n",
      "\n",
      "kmer=4\n",
      "token=250\n",
      "out_dict = \"result/20230808/\"\n",
      "out_dict_1 = \"result/20230808\"\n",
      "\n",
      "for virus_family in virus_l:\n",
      "    # 配列読み込み\n",
      "    df_seq = read_seq_f(virus_family, out_dict)\n",
      "    # seqを大文字に変換\n",
      "    df_seq[\"seq\"] = df_seq[\"seq\"].str.upper()\n",
      "    # kmerに分割\n",
      "    df_seq[\"sequence\"] = df_seq[\"seq\"].parallel_map(split_kmer)\n",
      "    # ラベル追加\n",
      "    df_token = add_infectivity_label(df_known, df_seq)\n",
      "    \n",
      "    # 5-cross fold val datasetの作成\n",
      "    fold_d, fold_index_d = five_cross_fold_val(df_known, virus_family)\n",
      "    \n",
      "    # ウイルス属ごとの配列数\n",
      "    df_fold, df_count = confirm_fold_data(df_known, virus_family, fold_index_d)\n",
      "    \n",
      "    # データセット分割状況の可視化\n",
      "    visualize_dataset(out_dict_1, virus_family, df_fold)\n",
      "    \n",
      "    # foldごとに出力\n",
      "    for fold, data in fold_d.items():\n",
      "        fold_name = \"fold_\" + str(fold)\n",
      "        \n",
      "        train_acc_l = data[0]\n",
      "        eval_acc_l = data[1]\n",
      "        test_acc_l = data[2]\n",
      "    \n",
      "        seq_d = {\"known_train\": [[out_dict_1, virus_family, \"ft_known_4_250_\" + fold_name, \"train.tsv\"], train_acc_l],\n",
      "             \"known_eval\": [[out_dict_1, virus_family, \"ft_known_4_250_\" + fold_name, \"dev.tsv\"], eval_acc_l],\n",
      "             \"known_pred\": [[out_dict_1, virus_family, \"pred_known_4_250_\" + fold_name, \"dev.tsv\"], test_acc_l]}\n",
      "\n",
      "        for key, data_1 in seq_d.items():\n",
      "            dict_l = data_1[0]\n",
      "            acc_l = data_1[1]\n",
      "            df = df_token[df_token[\"query\"].isin(acc_l)]\n",
      "            \n",
      "            df_output = df.rename(columns={\"host_label\":\"label\"})\n",
      "            df_output = df_output.loc[:, [\"sequence\", \"label\", \"seqid\"]]\n",
      "            df_output = df_output.astype({\"sequence\": str, \"label\": int, \"seqid\":str})\n",
      "\n",
      "            make_dir(dict_l[:-1])\n",
      "            output_f = \"/\".join(map(str, dict_l))\n",
      "\n",
      "            df_output.to_csv(output_f, sep=\"\\t\", index=False)\n",
      "            #print (df_output.shape)\n",
      "\n",
      "            # データセットの内訳\n",
      "            print_dataset_size(virus_family, key, fold, df_output)\n",
      "df_count\n",
      "%%bash\n",
      "\n",
      "virus=\"Arenaviridae\"\n",
      "\n",
      "for fold in \"fold_0\" \"fold_1\" \"fold_2\" \"fold_3\" \"fold_4\"\n",
      "do\n",
      "\n",
      "cat result/20230404/${virus}/ft_known_4_250_${fold}/train.tsv | wc -l\n",
      "cat result/20230718/${virus}/ft_known_4_250_${fold}/train.tsv | wc -l\n",
      "\n",
      "done\n",
      "%%bash\n",
      "\n",
      "for virus in \"Orthomyxoviridae\"  \"Hantaviridae\" \"Nairoviridae\" \"Peribunyaviridae\"\"Phenuiviridae\" \"Arenaviridae\" \"Reoviridae\"\n",
      "do\n",
      "for fold in \"fold_0\" \"fold_1\" \"fold_2\" \"fold_3\" \"fold_4\"\n",
      "do\n",
      "\n",
      "echo ${virus} ${fold} \"human_data\" \"animal_data\"\n",
      "\n",
      "awk -F'\\t' '$2 == 1 {count++} END {print count}' \\\n",
      "    result/20230404/${virus}/ft_known_4_250_${fold}/train.tsv \n",
      "\n",
      "awk -F'\\t' '$2 == 0 {count++} END {print count}' \\\n",
      "    result/20230404/${virus}/ft_known_4_250_${fold}/train.tsv\n",
      "\n",
      "done\n",
      "done\n",
      "%%bash\n",
      "\n",
      "for virus in \"Circoviridae\" \"Papillomaviridae\" \"Parvoviridae\" \"Polyomaviridae\" \\\n",
      "    \"Astroviridae\" \"Filoviridae\" \"Hepeviridae\" \"Pneumoviridae\"\n",
      "    \n",
      "do\n",
      "for fold in \"fold_0\" \"fold_1\" \"fold_2\" \"fold_3\" \"fold_4\"\n",
      "do\n",
      "\n",
      "echo ${virus} ${fold} \"human_data\" \"animal_data\"\n",
      "\n",
      "awk -F'\\t' '$2 == 1 {count++} END {print count}' \\\n",
      "    result/20230404/${virus}/ft_known_4_250_${fold}/train.tsv \n",
      "\n",
      "awk -F'\\t' '$2 == 0 {count++} END {print count}' \\\n",
      "    result/20230404/${virus}/ft_known_4_250_${fold}/train.tsv\n",
      "\n",
      "done\n",
      "done\n",
      "%history\n"
     ]
    }
   ],
   "source": [
    "%history"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env_210819",
   "language": "python",
   "name": "env_210819"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "197.797px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
